{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "75d60ade-4bbc-4e70-a48f-7b54df6cd37b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import os\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "afa7f279-9e82-41cb-8f7b-93a451119fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_half = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3d02b695-f951-4d05-9c74-d79fc593a2dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Dataset_Tag</th>\n",
       "      <th>Genome_ASM_PATH</th>\n",
       "      <th>ShortRead_Genome_ASM_PATH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N0072</td>\n",
       "      <td>ChinerOms_2019</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N0153</td>\n",
       "      <td>ChinerOms_2019</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TB3113</td>\n",
       "      <td>TB_Portals_24CI_R1</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TB1236</td>\n",
       "      <td>TB_Portals_24CI_R1</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TB2659</td>\n",
       "      <td>TB_Portals_24CI_R1</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/mm774/Projects/231121...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID         Dataset_Tag  \\\n",
       "0    N0072      ChinerOms_2019   \n",
       "1    N0153      ChinerOms_2019   \n",
       "2   TB3113  TB_Portals_24CI_R1   \n",
       "3   TB1236  TB_Portals_24CI_R1   \n",
       "4   TB2659  TB_Portals_24CI_R1   \n",
       "\n",
       "                                     Genome_ASM_PATH  \\\n",
       "0  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...   \n",
       "1  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...   \n",
       "2  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...   \n",
       "3  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...   \n",
       "4  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...   \n",
       "\n",
       "                           ShortRead_Genome_ASM_PATH  \n",
       "0  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...  \n",
       "1  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...  \n",
       "2  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...  \n",
       "3  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...  \n",
       "4  /n/data1/hms/dbmi/farhat/mm774/Projects/231121...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_paths_df = pd.read_csv(\"./231121.MtbSetV3.151CI.HybridAndSRAsm.FAPATHs.V1.tsv\", sep=\"\\t\")\n",
    "fasta_paths_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22fd4b8d-32ba-423c-909f-786581f20f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "l1_isolate = \"N0072\"\n",
    "l2_isolate = \"M0003941_3\"\n",
    "l3_isolate = \"N0004\"\n",
    "l4_isolate = \"M0017522_5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a101c3f0-126f-462d-9c19-2c90e1bc9b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "l1_fa_path = fasta_paths_df.loc[fasta_paths_df.SampleID == l1_isolate][\"Genome_ASM_PATH\"].values[0]\n",
    "l2_fa_path = fasta_paths_df.loc[fasta_paths_df.SampleID == l2_isolate][\"Genome_ASM_PATH\"].values[0]\n",
    "l3_fa_path = fasta_paths_df.loc[fasta_paths_df.SampleID == l3_isolate][\"Genome_ASM_PATH\"].values[0]\n",
    "l4_fa_path = fasta_paths_df.loc[fasta_paths_df.SampleID == l4_isolate][\"Genome_ASM_PATH\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8b654f5d-f1f1-4e19-9245-804726cc7dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "h37rv_fasta = \"/n/data1/hms/dbmi/farhat/mtb_data/h37rv/h37rv.fna\"\n",
    "with open(h37rv_fasta, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    h37rv_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        h37rv_seq += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4f39f8d9-df80-4f0c-88c1-1d5ff40713a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom_start</th>\n",
       "      <th>chrom_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_category_vargas2021</th>\n",
       "      <th>feature</th>\n",
       "      <th>function</th>\n",
       "      <th>product</th>\n",
       "      <th>functional_category</th>\n",
       "      <th>PFAM</th>\n",
       "      <th>gene_ontology</th>\n",
       "      <th>dr_mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dnaA</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>Essential</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Plays an important role in the initiation and ...</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "      <td>information pathways</td>\n",
       "      <td>P49993</td>\n",
       "      <td>GO:0003688,GO:0006270,GO:0005737,GO:0017111,GO...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dnaN</th>\n",
       "      <td>Rv0002</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>CDS</td>\n",
       "      <td>DNA polymerase III is a complex, multichain en...</td>\n",
       "      <td>DNA polymerase III (beta chain) DnaN (DNA nucl...</td>\n",
       "      <td>information pathways</td>\n",
       "      <td>Q50790</td>\n",
       "      <td>GO:0003677,GO:0006260,GO:0003887,GO:0005737,GO...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id  chrom_start  chrom_stop strand gene_category_vargas2021  \\\n",
       "gene_name                                                                    \n",
       "dnaA       Rv0001            1        1524      +                Essential   \n",
       "dnaN       Rv0002         2052        3260      +            Non-Essential   \n",
       "\n",
       "          feature                                           function  \\\n",
       "gene_name                                                              \n",
       "dnaA          CDS  Plays an important role in the initiation and ...   \n",
       "dnaN          CDS  DNA polymerase III is a complex, multichain en...   \n",
       "\n",
       "                                                     product  \\\n",
       "gene_name                                                      \n",
       "dnaA          Chromosomal replication initiator protein DnaA   \n",
       "dnaN       DNA polymerase III (beta chain) DnaN (DNA nucl...   \n",
       "\n",
       "            functional_category    PFAM  \\\n",
       "gene_name                                 \n",
       "dnaA       information pathways  P49993   \n",
       "dnaN       information pathways  Q50790   \n",
       "\n",
       "                                               gene_ontology dr_mutations  \n",
       "gene_name                                                                  \n",
       "dnaA       GO:0003688,GO:0006270,GO:0005737,GO:0017111,GO...          NaN  \n",
       "dnaN       GO:0003677,GO:0006260,GO:0003887,GO:0005737,GO...          NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h37rv_summary = pd.read_csv(\"/n/data1/hms/dbmi/farhat/mtb_data/mycobrowser/h37rv_mycobrowser_genome_summary.csv\").set_index(\"gene_name\")\n",
    "h37rv_summary.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43be3693-c8c6-435a-af2f-7110596078e8",
   "metadata": {},
   "source": [
    "# DR mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e117a026-8e4a-4668-8257-53a3593d82ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(513, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_locus</th>\n",
       "      <th>variant</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>nt_change</th>\n",
       "      <th>drug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_D461N</td>\n",
       "      <td>461.0</td>\n",
       "      <td>6620</td>\n",
       "      <td>A</td>\n",
       "      <td>LEVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_D461N</td>\n",
       "      <td>461.0</td>\n",
       "      <td>6620</td>\n",
       "      <td>A</td>\n",
       "      <td>MOXI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "      <td>499.0</td>\n",
       "      <td>6734</td>\n",
       "      <td>G</td>\n",
       "      <td>LEVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "      <td>499.0</td>\n",
       "      <td>6734</td>\n",
       "      <td>G</td>\n",
       "      <td>MOXI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_E501V</td>\n",
       "      <td>501.0</td>\n",
       "      <td>6741</td>\n",
       "      <td>T</td>\n",
       "      <td>LEVO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name gene_locus     variant  codon_number  genome_index nt_change  drug\n",
       "0      gyrB     Rv0005  gyrB_D461N         461.0          6620         A  LEVO\n",
       "1      gyrB     Rv0005  gyrB_D461N         461.0          6620         A  MOXI\n",
       "2      gyrB     Rv0005  gyrB_N499D         499.0          6734         G  LEVO\n",
       "3      gyrB     Rv0005  gyrB_N499D         499.0          6734         G  MOXI\n",
       "4      gyrB     Rv0005  gyrB_E501V         501.0          6741         T  LEVO"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dr_mutations = pd.read_csv(\"/home/sm624/data/antib_resistance/WHO_drug_SNP_mapping.csv\")\n",
    "print(dr_mutations.shape)\n",
    "dr_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9c771616-7533-4883-936d-dac805daa583",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "409"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_subst_ix_base_list = []\n",
    "\n",
    "for i in dr_mutations.index:\n",
    "    \n",
    "    pos_base = (dr_mutations.loc[i, \"genome_index\"]-1, dr_mutations.loc[i, \"nt_change\"])\n",
    "    \n",
    "    if pos_base not in DR_subst_ix_base_list:\n",
    "        DR_subst_ix_base_list.append(pos_base)\n",
    "        \n",
    "len(DR_subst_ix_base_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "660827da-6db4-42b1-9393-f98e2f2c2df2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(2288777, 'G'),\n",
       " (2289071, 'G'),\n",
       " (2289049, 'G'),\n",
       " (2288961, 'C'),\n",
       " (2289096, 'A'),\n",
       " (2288919, 'A'),\n",
       " (7569, 'T'),\n",
       " (2288816, 'T'),\n",
       " (4247468, 'G'),\n",
       " (4407795, 'A'),\n",
       " (761138, 'G'),\n",
       " (2288838, 'C'),\n",
       " (4327266, 'A'),\n",
       " (1473245, 'G'),\n",
       " (7581, 'C'),\n",
       " (2288972, 'C'),\n",
       " (2288886, 'C'),\n",
       " (2289104, 'T'),\n",
       " (761136, 'T'),\n",
       " (2289102, 'G')]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random.shuffle(DR_subst_ix_base_list)\n",
    "\n",
    "to_mutate_ix_base_list = []\n",
    "mutated_ix = []\n",
    "\n",
    "mutation_ix = 0\n",
    "\n",
    "while len(to_mutate_ix_base_list) < 20:\n",
    "\n",
    "    DR_mut_ix, DR_mut = DR_subst_ix_base_list[mutation_ix][0], DR_subst_ix_base_list[mutation_ix][1]\n",
    "\n",
    "    if not ((DR_mut_ix in mutated_ix) or (DR_mut_ix-1 in mutated_ix) or (DR_mut_ix+1 in mutated_ix)):\n",
    "\n",
    "        to_mutate_ix_base_list.append((DR_mut_ix, DR_mut))\n",
    "        mutated_ix.append(DR_mut_ix)\n",
    "        \n",
    "    mutation_ix += 1\n",
    "    \n",
    "to_mutate_ix_base_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "7a37b629-2d04-404d-a6f1-0b8b061074c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>mut_base</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGAAGTGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGAATAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGTAGAACACCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>A</td>\n",
       "      <td>GGAAGTCCTTGGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2288919</td>\n",
       "      <td>A</td>\n",
       "      <td>GACTCCTTCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7569</td>\n",
       "      <td>T</td>\n",
       "      <td>GCGACGCGTCGAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2288816</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGGCCGTCTGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4247468</td>\n",
       "      <td>G</td>\n",
       "      <td>CAACTATTTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4407795</td>\n",
       "      <td>A</td>\n",
       "      <td>CGCCCGTGACACCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>761138</td>\n",
       "      <td>G</td>\n",
       "      <td>TGACCCACAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2288838</td>\n",
       "      <td>C</td>\n",
       "      <td>GATCGGTGGCAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4327266</td>\n",
       "      <td>A</td>\n",
       "      <td>TCCGGTCCAGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1473245</td>\n",
       "      <td>G</td>\n",
       "      <td>CCGTCACGTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7581</td>\n",
       "      <td>C</td>\n",
       "      <td>TCTACGACAGCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2288972</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGCCTCGATTGCCGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2288886</td>\n",
       "      <td>C</td>\n",
       "      <td>GCAGCCAATTCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2289104</td>\n",
       "      <td>T</td>\n",
       "      <td>TTGGTTGCCACGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>761136</td>\n",
       "      <td>T</td>\n",
       "      <td>GGTTGACCCACAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2289102</td>\n",
       "      <td>G</td>\n",
       "      <td>CCTTGGTTGCCAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    h37rv_ix mut_base               kmer\n",
       "0    2288777        G        CCAGCACCCTG\n",
       "1    2289071        G      GAGAAGTGGTCAC\n",
       "2    2289049        G        GAGGAATAGTC\n",
       "3    2288961        C    TTGTAGAACACCGCC\n",
       "4    2289096        A      GGAAGTCCTTGGT\n",
       "5    2288919        A        GACTCCTTCGA\n",
       "6       7569        T      GCGACGCGTCGAT\n",
       "7    2288816        T      TCGGCCGTCTGGC\n",
       "8    4247468        G        CAACTATTTCC\n",
       "9    4407795        A    CGCCCGTGACACCGC\n",
       "10    761138        G        TGACCCACAAG\n",
       "11   2288838        C      GATCGGTGGCAAT\n",
       "12   4327266        A      TCCGGTCCAGGGA\n",
       "13   1473245        G        CCGTCACGTCA\n",
       "14      7581        C      TCTACGACAGCCT\n",
       "15   2288972        C  CCGCCTCGATTGCCGAC\n",
       "16   2288886        C      GCAGCCAATTCAG\n",
       "17   2289104        T      TTGGTTGCCACGA\n",
       "18    761136        T      GGTTGACCCACAA\n",
       "19   2289102        G      CCTTGGTTGCCAC"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_mut_kmers_df = pd.DataFrame(columns=[\"h37rv_ix\", \"mut_base\", \"kmer\"])\n",
    "df_i = 0\n",
    "\n",
    "for ix, base in to_mutate_ix_base_list:\n",
    "    \n",
    "    h37rv_ref_base = h37rv_seq[ix]\n",
    "    \n",
    "    h37rv_kmer = h37rv_seq[ix-kmer_half:ix+kmer_half+1]\n",
    "    \n",
    "    new_kmer_half = kmer_half\n",
    "    while h37rv_seq.count(h37rv_kmer) > 1:\n",
    "        new_kmer_half += 1\n",
    "        h37rv_kmer = h37rv_seq[ix-new_kmer_half:ix+new_kmer_half+1]\n",
    "    \n",
    "    DR_mut_kmers_df.loc[df_i] = [ix, base, h37rv_kmer]\n",
    "    df_i += 1\n",
    "    \n",
    "DR_mut_kmers_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "40d5cdd4-e00d-45ed-94cc-2cdf7a9789a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "DR_mut_kmers_df.to_csv(\"./DR_mut_h37rv_kmers.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3096722a-e4cf-45ee-95a5-e9c7469be9ba",
   "metadata": {},
   "source": [
    "## Check L1 kmer matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "abc0280c-0429-4be7-b408-04abf514f4b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>mut_base</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGAAGTGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGAATAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGTAGAACACCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>A</td>\n",
       "      <td>GGAAGTCCTTGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix mut_base             kmer\n",
       "0   2288777        G      CCAGCACCCTG\n",
       "1   2289071        G    GAGAAGTGGTCAC\n",
       "2   2289049        G      GAGGAATAGTC\n",
       "3   2288961        C  TTGTAGAACACCGCC\n",
       "4   2289096        A    GGAAGTCCTTGGT"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_mut_kmers_df = pd.read_csv(\"./DR_mut_h37rv_kmers.csv\")\n",
    "DR_mut_kmers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7c03d2d4-23f9-433d-ae3a-30617c64530b",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l1_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L1_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L1_seq += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ce66c2a-4a5a-481a-a8ac-18e853901c93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4421404, 4411532)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(L1_seq), len(h37rv_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aa596e93-b985-4e91-b9fb-9dbe31717f22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h37rv_ix             7581\n",
      "mut_base                C\n",
      "kmer        TCTACGACAGCCT\n",
      "Name: 14, dtype: object\n"
     ]
    }
   ],
   "source": [
    "for i in DR_mut_kmers_df.index:\n",
    "    \n",
    "    kmer = DR_mut_kmers_df.loc[i, \"kmer\"]\n",
    "    \n",
    "    if L1_seq.count(kmer) != 1:\n",
    "        print(DR_mut_kmers_df.loc[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "910b16cc-8112-4696-8f04-cd52d4483049",
   "metadata": {},
   "source": [
    "## Check L2 kmer matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2579e4cb-fbf0-4dce-ad15-57fac0f68c69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>mut_base</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGAAGTGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGAATAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGTAGAACACCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>A</td>\n",
       "      <td>GGAAGTCCTTGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix mut_base             kmer\n",
       "0   2288777        G      CCAGCACCCTG\n",
       "1   2289071        G    GAGAAGTGGTCAC\n",
       "2   2289049        G      GAGGAATAGTC\n",
       "3   2288961        C  TTGTAGAACACCGCC\n",
       "4   2289096        A    GGAAGTCCTTGGT"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_mut_kmers_df = pd.read_csv(\"./DR_mut_h37rv_kmers.csv\")\n",
    "DR_mut_kmers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8eabb367-1b6c-4b83-85d4-267b729cdc9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l2_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L2_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L2_seq += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c56e47e3-faa9-4a37-8a98-a0fbd4c3ed4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4421169, 4411532)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(L2_seq), len(h37rv_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c08c686f-d5eb-4d9e-8df7-6891f2476c05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h37rv_ix             7581\n",
      "mut_base                C\n",
      "kmer        TCTACGACAGCCT\n",
      "Name: 14, dtype: object\n"
     ]
    }
   ],
   "source": [
    "for i in DR_mut_kmers_df.index:\n",
    "    \n",
    "    kmer = DR_mut_kmers_df.loc[i, \"kmer\"]\n",
    "    \n",
    "    if L2_seq.count(kmer) != 1:\n",
    "        print(DR_mut_kmers_df.loc[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed1be60-6557-4a76-882b-3a2582554cdf",
   "metadata": {},
   "source": [
    "## Check L3 kmer matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d603523d-4cd8-45e8-b3a4-45a14287b059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>mut_base</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGAAGTGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGAATAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGTAGAACACCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>A</td>\n",
       "      <td>GGAAGTCCTTGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix mut_base             kmer\n",
       "0   2288777        G      CCAGCACCCTG\n",
       "1   2289071        G    GAGAAGTGGTCAC\n",
       "2   2289049        G      GAGGAATAGTC\n",
       "3   2288961        C  TTGTAGAACACCGCC\n",
       "4   2289096        A    GGAAGTCCTTGGT"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_mut_kmers_df = pd.read_csv(\"./DR_mut_h37rv_kmers.csv\")\n",
    "DR_mut_kmers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f6e73aec-2024-4c13-a08a-5e82862bc819",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l3_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L3_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L3_seq += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "78f8b4fe-ffd6-4499-bb63-b2186beb1ecf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4423516, 4411532)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(L3_seq), len(h37rv_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d97e30d3-aac1-4d6e-bcc9-4e0405153e4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h37rv_ix        2289049\n",
      "mut_base              G\n",
      "kmer        GAGGAATAGTC\n",
      "Name: 2, dtype: object\n",
      "h37rv_ix             7581\n",
      "mut_base                C\n",
      "kmer        TCTACGACAGCCT\n",
      "Name: 14, dtype: object\n"
     ]
    }
   ],
   "source": [
    "for i in DR_mut_kmers_df.index:\n",
    "    \n",
    "    kmer = DR_mut_kmers_df.loc[i, \"kmer\"]\n",
    "    \n",
    "    if L3_seq.count(kmer) != 1:\n",
    "        print(DR_mut_kmers_df.loc[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82fcb01a-cc0a-454d-9b19-7e59131c7c35",
   "metadata": {},
   "source": [
    "## Check L4 kmer matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ddc26796-a63f-4909-819a-b10adc3bca01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>mut_base</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGAAGTGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGAATAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGTAGAACACCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>A</td>\n",
       "      <td>GGAAGTCCTTGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix mut_base             kmer\n",
       "0   2288777        G      CCAGCACCCTG\n",
       "1   2289071        G    GAGAAGTGGTCAC\n",
       "2   2289049        G      GAGGAATAGTC\n",
       "3   2288961        C  TTGTAGAACACCGCC\n",
       "4   2289096        A    GGAAGTCCTTGGT"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_mut_kmers_df = pd.read_csv(\"./DR_mut_h37rv_kmers.csv\")\n",
    "DR_mut_kmers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5594fcf0-24e0-44f1-bb62-38fb9334ebbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l4_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L4_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L4_seq += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "663115d8-5f9b-48c2-934b-c467be6ed2bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4416715, 4411532)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(L4_seq), len(h37rv_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "047fc6b3-95cd-4056-8822-2d55f08ab66b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h37rv_ix             7581\n",
      "mut_base                C\n",
      "kmer        TCTACGACAGCCT\n",
      "Name: 14, dtype: object\n"
     ]
    }
   ],
   "source": [
    "for i in DR_mut_kmers_df.index:\n",
    "    \n",
    "    kmer = DR_mut_kmers_df.loc[i, \"kmer\"]\n",
    "    \n",
    "    if L4_seq.count(kmer) != 1:\n",
    "        print(DR_mut_kmers_df.loc[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d632d2bf-f4f3-403b-b32f-5e01d94d452f",
   "metadata": {},
   "source": [
    "## BLAST analysis of problematic positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6e714cd2-856e-49b5-8c74-a82ea0bf2a8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_locus</th>\n",
       "      <th>variant</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>nt_change</th>\n",
       "      <th>drug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>gyrA</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>7582</td>\n",
       "      <td>G</td>\n",
       "      <td>LEVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>gyrA</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>7582</td>\n",
       "      <td>G</td>\n",
       "      <td>MOXI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gyrA</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA_D94A</td>\n",
       "      <td>94.0</td>\n",
       "      <td>7582</td>\n",
       "      <td>C</td>\n",
       "      <td>LEVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>gyrA</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA_D94A</td>\n",
       "      <td>94.0</td>\n",
       "      <td>7582</td>\n",
       "      <td>C</td>\n",
       "      <td>MOXI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name gene_locus    variant  codon_number  genome_index nt_change  drug\n",
       "24      gyrA     Rv0006  gyrA_D94G          94.0          7582         G  LEVO\n",
       "25      gyrA     Rv0006  gyrA_D94G          94.0          7582         G  MOXI\n",
       "26      gyrA     Rv0006  gyrA_D94A          94.0          7582         C  LEVO\n",
       "27      gyrA     Rv0006  gyrA_D94A          94.0          7582         C  MOXI"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dr_mutations[dr_mutations.genome_index == 7582]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "efd3b0ce-1ffc-4db9-896c-3fd0ffcc46cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_locus</th>\n",
       "      <th>variant</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>nt_change</th>\n",
       "      <th>drug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>pncA</td>\n",
       "      <td>Rv2043c</td>\n",
       "      <td>pncA_Y64!</td>\n",
       "      <td>64.0</td>\n",
       "      <td>2289050</td>\n",
       "      <td>G</td>\n",
       "      <td>PZA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name gene_locus    variant  codon_number  genome_index nt_change drug\n",
       "306      pncA    Rv2043c  pncA_Y64!          64.0       2289050         G  PZA"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dr_mutations[dr_mutations.genome_index == 2289050]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48ba0f3a-f210-41b3-b19e-eac294a80551",
   "metadata": {},
   "source": [
    "### Check nearby lineage markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "6bb509bf-e45e-4cef-8512-691842bc495f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>position</th>\n",
       "      <th>gene_coordinate</th>\n",
       "      <th>allele_change</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>amino_acid_change</th>\n",
       "      <th>locus_Id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>essential</th>\n",
       "      <th>diagnostic_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lineage1</td>\n",
       "      <td>6112</td>\n",
       "      <td>873</td>\n",
       "      <td>G/C</td>\n",
       "      <td>291</td>\n",
       "      <td>ATG/ATC</td>\n",
       "      <td>M/I</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>essential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lineage1</td>\n",
       "      <td>8452</td>\n",
       "      <td>1151</td>\n",
       "      <td>C/T</td>\n",
       "      <td>384</td>\n",
       "      <td>GCA/GTA</td>\n",
       "      <td>A/V</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lineage1</td>\n",
       "      <td>13298</td>\n",
       "      <td>261</td>\n",
       "      <td>G/C</td>\n",
       "      <td>87</td>\n",
       "      <td>ATC/ATG</td>\n",
       "      <td>I/M</td>\n",
       "      <td>Rv0010c</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lineage1</td>\n",
       "      <td>22961</td>\n",
       "      <td>221</td>\n",
       "      <td>T/G</td>\n",
       "      <td>74</td>\n",
       "      <td>GAT/GCT</td>\n",
       "      <td>D/A</td>\n",
       "      <td>Rv0018c</td>\n",
       "      <td>pstP</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lineage1</td>\n",
       "      <td>26347</td>\n",
       "      <td>535</td>\n",
       "      <td>C/G</td>\n",
       "      <td>179</td>\n",
       "      <td>GAC/CAC</td>\n",
       "      <td>D/H</td>\n",
       "      <td>Rv0021c</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6910</th>\n",
       "      <td>lineageBOV_AFRI</td>\n",
       "      <td>4292095</td>\n",
       "      <td>1080</td>\n",
       "      <td>C/A</td>\n",
       "      <td>360</td>\n",
       "      <td>GCG/GCT</td>\n",
       "      <td>A/A</td>\n",
       "      <td>Rv3824c</td>\n",
       "      <td>papA1</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6911</th>\n",
       "      <td>lineageBOV_AFRI</td>\n",
       "      <td>4330238</td>\n",
       "      <td>1268</td>\n",
       "      <td>C/T</td>\n",
       "      <td>423</td>\n",
       "      <td>CGT/CAT</td>\n",
       "      <td>R/H</td>\n",
       "      <td>Rv3858c</td>\n",
       "      <td>gltD</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>essential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6912</th>\n",
       "      <td>lineageBOV_AFRI</td>\n",
       "      <td>4386257</td>\n",
       "      <td>52</td>\n",
       "      <td>G/C</td>\n",
       "      <td>18</td>\n",
       "      <td>CCG/GCG</td>\n",
       "      <td>P/A</td>\n",
       "      <td>Rv3900c</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6913</th>\n",
       "      <td>lineageBOV_AFRI</td>\n",
       "      <td>4393838</td>\n",
       "      <td>390</td>\n",
       "      <td>C/T</td>\n",
       "      <td>130</td>\n",
       "      <td>CTC/CTT</td>\n",
       "      <td>L/L</td>\n",
       "      <td>Rv3908</td>\n",
       "      <td>mutT4</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6914</th>\n",
       "      <td>lineageBOV_AFRI</td>\n",
       "      <td>4408570</td>\n",
       "      <td>328</td>\n",
       "      <td>G/A</td>\n",
       "      <td>110</td>\n",
       "      <td>CGG/TGG</td>\n",
       "      <td>R/W</td>\n",
       "      <td>Rv3920c</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6915 rows  12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              lineage  position  gene_coordinate allele_change  codon_number  \\\n",
       "0            lineage1      6112              873           G/C           291   \n",
       "1            lineage1      8452             1151           C/T           384   \n",
       "2            lineage1     13298              261           G/C            87   \n",
       "3            lineage1     22961              221           T/G            74   \n",
       "4            lineage1     26347              535           C/G           179   \n",
       "...               ...       ...              ...           ...           ...   \n",
       "6910  lineageBOV_AFRI   4292095             1080           C/A           360   \n",
       "6911  lineageBOV_AFRI   4330238             1268           C/T           423   \n",
       "6912  lineageBOV_AFRI   4386257               52           G/C            18   \n",
       "6913  lineageBOV_AFRI   4393838              390           C/T           130   \n",
       "6914  lineageBOV_AFRI   4408570              328           G/A           110   \n",
       "\n",
       "     codon_change amino_acid_change locus_Id gene_name mutation_type  \\\n",
       "0         ATG/ATC               M/I   Rv0005      gyrB        nonsyn   \n",
       "1         GCA/GTA               A/V   Rv0006      gyrA        nonsyn   \n",
       "2         ATC/ATG               I/M  Rv0010c         0        nonsyn   \n",
       "3         GAT/GCT               D/A  Rv0018c      pstP        nonsyn   \n",
       "4         GAC/CAC               D/H  Rv0021c         0        nonsyn   \n",
       "...           ...               ...      ...       ...           ...   \n",
       "6910      GCG/GCT               A/A  Rv3824c     papA1           syn   \n",
       "6911      CGT/CAT               R/H  Rv3858c      gltD        nonsyn   \n",
       "6912      CCG/GCG               P/A  Rv3900c         0        nonsyn   \n",
       "6913      CTC/CTT               L/L   Rv3908     mutT4           syn   \n",
       "6914      CGG/TGG               R/W  Rv3920c         0        nonsyn   \n",
       "\n",
       "         essential diagnostic_snp  \n",
       "0        essential             no  \n",
       "1     nonessential             no  \n",
       "2     nonessential             no  \n",
       "3     nonessential             no  \n",
       "4     nonessential             no  \n",
       "...            ...            ...  \n",
       "6910  nonessential             no  \n",
       "6911     essential             no  \n",
       "6912  nonessential             no  \n",
       "6913  nonessential             no  \n",
       "6914  nonessential             no  \n",
       "\n",
       "[6915 rows x 12 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lineage_snps = pd.read_csv(\"/home/sm624/projects/calculate_F2/Coll2014_LinSpeSNPs_final.csv\")\n",
    "lineage_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "506a904c-ef96-4371-beeb-bdb8d4d519d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>position</th>\n",
       "      <th>gene_coordinate</th>\n",
       "      <th>allele_change</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>amino_acid_change</th>\n",
       "      <th>locus_Id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>essential</th>\n",
       "      <th>diagnostic_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lineage1</td>\n",
       "      <td>8452</td>\n",
       "      <td>1151</td>\n",
       "      <td>C/T</td>\n",
       "      <td>384</td>\n",
       "      <td>GCA/GTA</td>\n",
       "      <td>A/V</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>931</th>\n",
       "      <td>lineage1.2.1</td>\n",
       "      <td>9260</td>\n",
       "      <td>1959</td>\n",
       "      <td>G/C</td>\n",
       "      <td>653</td>\n",
       "      <td>CTG/CTC</td>\n",
       "      <td>L/L</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2180</th>\n",
       "      <td>lineage3.1.2.2</td>\n",
       "      <td>9611</td>\n",
       "      <td>2310</td>\n",
       "      <td>C/T</td>\n",
       "      <td>770</td>\n",
       "      <td>GAC/GAT</td>\n",
       "      <td>D/D</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3650</th>\n",
       "      <td>lineage4.3.3</td>\n",
       "      <td>8040</td>\n",
       "      <td>739</td>\n",
       "      <td>G/A</td>\n",
       "      <td>247</td>\n",
       "      <td>GGC/AGC</td>\n",
       "      <td>G/S</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4263</th>\n",
       "      <td>lineage4.5</td>\n",
       "      <td>7892</td>\n",
       "      <td>591</td>\n",
       "      <td>G/A</td>\n",
       "      <td>197</td>\n",
       "      <td>CTG/CTA</td>\n",
       "      <td>L/L</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4422</th>\n",
       "      <td>lineage4.6.1</td>\n",
       "      <td>7539</td>\n",
       "      <td>238</td>\n",
       "      <td>A/G</td>\n",
       "      <td>80</td>\n",
       "      <td>ACC/GCC</td>\n",
       "      <td>T/A</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>nonsyn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5211</th>\n",
       "      <td>lineage5</td>\n",
       "      <td>9566</td>\n",
       "      <td>2265</td>\n",
       "      <td>C/T</td>\n",
       "      <td>755</td>\n",
       "      <td>TAC/TAT</td>\n",
       "      <td>Y/Y</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5804</th>\n",
       "      <td>lineage7</td>\n",
       "      <td>8876</td>\n",
       "      <td>1575</td>\n",
       "      <td>C/T</td>\n",
       "      <td>525</td>\n",
       "      <td>TAC/TAT</td>\n",
       "      <td>Y/Y</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>syn</td>\n",
       "      <td>nonessential</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             lineage  position  gene_coordinate allele_change  codon_number  \\\n",
       "1           lineage1      8452             1151           C/T           384   \n",
       "931     lineage1.2.1      9260             1959           G/C           653   \n",
       "2180  lineage3.1.2.2      9611             2310           C/T           770   \n",
       "3650    lineage4.3.3      8040              739           G/A           247   \n",
       "4263      lineage4.5      7892              591           G/A           197   \n",
       "4422    lineage4.6.1      7539              238           A/G            80   \n",
       "5211        lineage5      9566             2265           C/T           755   \n",
       "5804        lineage7      8876             1575           C/T           525   \n",
       "\n",
       "     codon_change amino_acid_change locus_Id gene_name mutation_type  \\\n",
       "1         GCA/GTA               A/V   Rv0006      gyrA        nonsyn   \n",
       "931       CTG/CTC               L/L   Rv0006      gyrA           syn   \n",
       "2180      GAC/GAT               D/D   Rv0006      gyrA           syn   \n",
       "3650      GGC/AGC               G/S   Rv0006      gyrA        nonsyn   \n",
       "4263      CTG/CTA               L/L   Rv0006      gyrA           syn   \n",
       "4422      ACC/GCC               T/A   Rv0006      gyrA        nonsyn   \n",
       "5211      TAC/TAT               Y/Y   Rv0006      gyrA           syn   \n",
       "5804      TAC/TAT               Y/Y   Rv0006      gyrA           syn   \n",
       "\n",
       "         essential diagnostic_snp  \n",
       "1     nonessential             no  \n",
       "931   nonessential             no  \n",
       "2180  nonessential             no  \n",
       "3650  nonessential             no  \n",
       "4263  nonessential             no  \n",
       "4422  nonessential             no  \n",
       "5211  nonessential             no  \n",
       "5804  nonessential             no  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lineage_snps[lineage_snps.gene_name == \"gyrA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "aaae71a3-f491-4173-88b8-9763b4be3702",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>position</th>\n",
       "      <th>gene_coordinate</th>\n",
       "      <th>allele_change</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>amino_acid_change</th>\n",
       "      <th>locus_Id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>essential</th>\n",
       "      <th>diagnostic_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lineage, position, gene_coordinate, allele_change, codon_number, codon_change, amino_acid_change, locus_Id, gene_name, mutation_type, essential, diagnostic_snp]\n",
       "Index: []"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lineage_snps[lineage_snps.gene_name == \"pncA\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f1592fd-97fc-460a-876f-88ae11d338f2",
   "metadata": {},
   "source": [
    "### Create gene fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6ff0f17b-8fc7-42f6-8fd8-e05ffb2f2427",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"gyrA\", \"pncA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "189370a4-245e-4911-a231-56a60dfd734c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom_start</th>\n",
       "      <th>chrom_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_category_vargas2021</th>\n",
       "      <th>feature</th>\n",
       "      <th>function</th>\n",
       "      <th>product</th>\n",
       "      <th>functional_category</th>\n",
       "      <th>PFAM</th>\n",
       "      <th>gene_ontology</th>\n",
       "      <th>dr_mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dnaA</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>Essential</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Plays an important role in the initiation and ...</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "      <td>information pathways</td>\n",
       "      <td>P49993</td>\n",
       "      <td>GO:0003688,GO:0006270,GO:0005737,GO:0017111,GO...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dnaN</th>\n",
       "      <td>Rv0002</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>CDS</td>\n",
       "      <td>DNA polymerase III is a complex, multichain en...</td>\n",
       "      <td>DNA polymerase III (beta chain) DnaN (DNA nucl...</td>\n",
       "      <td>information pathways</td>\n",
       "      <td>Q50790</td>\n",
       "      <td>GO:0003677,GO:0006260,GO:0003887,GO:0005737,GO...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id  chrom_start  chrom_stop strand gene_category_vargas2021  \\\n",
       "gene_name                                                                    \n",
       "dnaA       Rv0001            1        1524      +                Essential   \n",
       "dnaN       Rv0002         2052        3260      +            Non-Essential   \n",
       "\n",
       "          feature                                           function  \\\n",
       "gene_name                                                              \n",
       "dnaA          CDS  Plays an important role in the initiation and ...   \n",
       "dnaN          CDS  DNA polymerase III is a complex, multichain en...   \n",
       "\n",
       "                                                     product  \\\n",
       "gene_name                                                      \n",
       "dnaA          Chromosomal replication initiator protein DnaA   \n",
       "dnaN       DNA polymerase III (beta chain) DnaN (DNA nucl...   \n",
       "\n",
       "            functional_category    PFAM  \\\n",
       "gene_name                                 \n",
       "dnaA       information pathways  P49993   \n",
       "dnaN       information pathways  Q50790   \n",
       "\n",
       "                                               gene_ontology dr_mutations  \n",
       "gene_name                                                                  \n",
       "dnaA       GO:0003688,GO:0006270,GO:0005737,GO:0017111,GO...          NaN  \n",
       "dnaN       GO:0003677,GO:0006260,GO:0003887,GO:0005737,GO...          NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h37rv_summary = pd.read_csv(\"/n/data1/hms/dbmi/farhat/mtb_data/mycobrowser/h37rv_mycobrowser_genome_summary.csv\").set_index(\"gene_name\")\n",
    "h37rv_summary.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "cc7a8a39-45bb-4220-8e65-e03bd2b1f9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chrom_start and chrom_stop are inclusive\n",
    "\n",
    "for gene in genes:\n",
    "    \n",
    "    chrom_start, chrom_stop = h37rv_summary.loc[gene, \"chrom_start\"], h37rv_summary.loc[gene, \"chrom_stop\"]\n",
    "    \n",
    "    gene_seq = h37rv_seq[chrom_start-1:chrom_stop]\n",
    "    \n",
    "    with open(\"./blast/{}.fa\".format(gene), \"w\") as out_f:\n",
    "        out_f.write(\">{}\\n\".format(gene))\n",
    "        n_chunks = len(gene_seq)//80\n",
    "        for i in range(n_chunks):\n",
    "            out_f.write(gene_seq[i*80:i*80+80] + \"\\n\")\n",
    "        out_f.write(gene_seq[(i+1)*80:] + \"\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b55681c-3cf2-422e-85b3-62027209846d",
   "metadata": {},
   "source": [
    "```\n",
    "makeblastdb -in gyrA.fa -dbtype nucl\n",
    "\n",
    "makeblastdb -in pncA.fa -dbtype nucl\n",
    "```\n",
    "\n",
    "```\n",
    "## gyrA\n",
    "### lineage 1\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/ChinerOms_2019/N0072.LR.Asm.fasta -db gyrA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage1.gyrA.blastn.txt\n",
    "\n",
    "### lineage 2\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/Farhat_Peru_2019/M0003941_3.LR.Asm.fasta -db gyrA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage2.gyrA.blastn.txt\n",
    "\n",
    "### lineage 3\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/ChinerOms_2019/N0004.LR.Asm.fasta -db gyrA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage3.gyrA.blastn.txt\n",
    "\n",
    "### lineage 4\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/Farhat_Peru_2019/M0017522_5.LR.Asm.fasta -db gyrA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage4.gyrA.blastn.txt\n",
    "```\n",
    "\n",
    "```\n",
    "## pncA\n",
    "### lineage 1\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/ChinerOms_2019/N0072.LR.Asm.fasta -db pncA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage1.pncA.blastn.txt\n",
    "\n",
    "### lineage 2\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/Farhat_Peru_2019/M0003941_3.LR.Asm.fasta -db pncA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage2.pncA.blastn.txt\n",
    "\n",
    "### lineage 3\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/ChinerOms_2019/N0004.LR.Asm.fasta -db pncA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage3.pncA.blastn.txt\n",
    "\n",
    "### lineage 4\n",
    "blastn -query /n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/Farhat_Peru_2019/M0017522_5.LR.Asm.fasta -db pncA.fa -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq nident\" -out /home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage4.pncA.blastn.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bc0c010-3860-4c55-878d-ef35d3ffdad5",
   "metadata": {},
   "source": [
    "### Analyze BLAST results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "20ef1a8b-501d-4b58-b4db-e89a9b4d3df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"gyrA\", \"pncA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "64676fd3-8699-4c27-82f3-3aff3d96cf72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4604</td>\n",
       "      <td>2517</td>\n",
       "      <td>ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      4604    2517  ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1037</td>\n",
       "      <td>561</td>\n",
       "      <td>TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      1037     561  TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4626</td>\n",
       "      <td>2517</td>\n",
       "      <td>ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      4626    2517  ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1037</td>\n",
       "      <td>561</td>\n",
       "      <td>TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      1037     561  TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4632</td>\n",
       "      <td>2517</td>\n",
       "      <td>ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      4632    2517  ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1031</td>\n",
       "      <td>561</td>\n",
       "      <td>TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      1031     561  TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4632</td>\n",
       "      <td>2517</td>\n",
       "      <td>ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      4632    2517  ATGACAGACACGACGTTGCCGCCTGACGACTCGCTCGACCGGATCG..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>qseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1037</td>\n",
       "      <td>561</td>\n",
       "      <td>TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   evalue  bitscore  length                                               qseq\n",
       "0     0.0      1037     561  TCAGGAGCTGCAAACCAACTCGACGCTGGCGGTGCGCATCTCCTCC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for lin_i in range(1, 5):\n",
    "    for gene in genes:\n",
    "        df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage{}.{}.blastn.txt\".format(lin_i, gene),\n",
    "                         sep=\"\\t\",\n",
    "                         names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "        display(df[[\"evalue\", \"bitscore\", \"length\", \"qseq\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cc879a89-7587-4393-9808-c517bf872ab0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lineage 1 >> 7581 CTACGACAGCC CTACGACACCC\n",
      "lineage 2 >> 7581 CTACGACAGCC CTACGACACCC\n",
      "lineage 3 >> 7581 CTACGACAGCC CTACGACACCC\n",
      "lineage 4 >> 7581 CTACGACAGCC CTACGACACCC\n"
     ]
    }
   ],
   "source": [
    "# 7581 gyrA\n",
    "# chrom_start and chrom_stop are inclusive and 1 indexed\n",
    "\n",
    "mutate_ix = 7581\n",
    "\n",
    "h37rv_ref_base = h37rv_seq[mutate_ix-5:mutate_ix+6]\n",
    "\n",
    "chrom_start, chrom_stop = h37rv_summary.loc[\"gyrA\", \"chrom_start\"], h37rv_summary.loc[\"gyrA\", \"chrom_stop\"]\n",
    "\n",
    "gene_ix = mutate_ix - chrom_start + 1\n",
    "\n",
    "l1_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage1.gyrA.blastn.txt\",\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l1_gyrA_seq = l1_df.qseq.values[0]\n",
    "l2_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage2.gyrA.blastn.txt\",\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l2_gyrA_seq = l2_df.qseq.values[0]\n",
    "l3_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage3.gyrA.blastn.txt\",\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l3_gyrA_seq = l3_df.qseq.values[0]\n",
    "l4_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage4.gyrA.blastn.txt\",\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l4_gyrA_seq = l4_df.qseq.values[0]\n",
    "\n",
    "print(\"lineage 1 >>\", mutate_ix, h37rv_ref_base, l1_gyrA_seq[gene_ix-5:gene_ix+6])\n",
    "print(\"lineage 2 >>\", mutate_ix, h37rv_ref_base, l2_gyrA_seq[gene_ix-5:gene_ix+6])\n",
    "print(\"lineage 3 >>\", mutate_ix, h37rv_ref_base, l3_gyrA_seq[gene_ix-5:gene_ix+6])\n",
    "print(\"lineage 4 >>\", mutate_ix, h37rv_ref_base, l4_gyrA_seq[gene_ix-5:gene_ix+6])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c02a1b82-b430-4acd-b1bf-cd14110137d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ATGACAGACA', 'ATGACAGACA')"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gyrA_seq[0:10], l1_gyrA_seq[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3f83eb0a-56cc-4c5f-afbc-a53323fc84a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    8\n",
       " Name: mismatch, dtype: int64,\n",
       " 0    4\n",
       " Name: mismatch, dtype: int64,\n",
       " 0    3\n",
       " Name: mismatch, dtype: int64,\n",
       " 0    3\n",
       " Name: mismatch, dtype: int64)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l1_df.mismatch, l2_df.mismatch, l3_df.mismatch, l4_df.mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e5633b-64b5-408a-a5b6-b60708fe1ef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "l1_gyrA_seq == gyrA_seq, l2_gyrA_seq == gyrA_seq, l3_gyrA_seq == gyrA_seq, l4_gyrA_seq == gyrA_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "9d8bce32-3e1c-40ff-a50a-57959fc8b56b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60 283 669 1150 1745 1841 2002 2141 \n"
     ]
    }
   ],
   "source": [
    "gyrA_seq = h37rv_seq[h37rv_summary.loc[\"gyrA\", \"chrom_start\"]-1:h37rv_summary.loc[\"gyrA\", \"chrom_stop\"]]\n",
    "\n",
    "for i in range(len(l1_gyrA_seq)):\n",
    "    if gyrA_seq[i] != l1_gyrA_seq[i]:\n",
    "        print(i, end=\" \")\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d2ca75b8-7c83-4c98-a265-e77b6184dcd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lineage 1 >> 2289049 GAGGAATAGTC GAGGAATAGTC\n",
      "lineage 2 >> 2289049 GAGGAATAGTC GAGGAATAGTC\n",
      "lineage 3 >> 2289049 GAGGAATAGTC GAAGAATAGTC\n",
      "lineage 4 >> 2289049 GAGGAATAGTC GAGGAATAGTC\n"
     ]
    }
   ],
   "source": [
    "# 2289049 pncA\n",
    "# chrom_start and chrom_stop are inclusive and 1 indexed\n",
    "\n",
    "mutate_ix = 2289049\n",
    "\n",
    "h37rv_ref_base = h37rv_seq[mutate_ix-5:mutate_ix+6]\n",
    "\n",
    "chrom_start, chrom_stop = h37rv_summary.loc[\"pncA\", \"chrom_start\"], h37rv_summary.loc[\"pncA\", \"chrom_stop\"]\n",
    "\n",
    "gene_ix = mutate_ix - chrom_start + 1\n",
    "\n",
    "l1_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage1.pncA.blastn.txt\".format(lin_i, gene),\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l1_pncA_seq = l1_df.qseq.values[0]\n",
    "l2_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage2.pncA.blastn.txt\".format(lin_i, gene),\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l2_pncA_seq = l2_df.qseq.values[0]\n",
    "l3_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage3.pncA.blastn.txt\".format(lin_i, gene),\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l3_pncA_seq = l3_df.qseq.values[0]\n",
    "l4_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage4.pncA.blastn.txt\".format(lin_i, gene),\n",
    "                 sep=\"\\t\",\n",
    "                 names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "l4_pncA_seq = l4_df.qseq.values[0]\n",
    "\n",
    "print(\"lineage 1 >>\", mutate_ix, h37rv_ref_base, l1_pncA_seq[gene_ix-5:gene_ix+6])\n",
    "print(\"lineage 2 >>\", mutate_ix, h37rv_ref_base, l2_pncA_seq[gene_ix-5:gene_ix+6])\n",
    "print(\"lineage 3 >>\", mutate_ix, h37rv_ref_base, l3_pncA_seq[gene_ix-5:gene_ix+6])\n",
    "print(\"lineage 4 >>\", mutate_ix, h37rv_ref_base, l4_pncA_seq[gene_ix-5:gene_ix+6])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f93962aa-b39c-448e-8abb-c7aa8d0a7ba2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1\n",
       "Name: mismatch, dtype: int64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l3_df.mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "05f87d58-a203-4d1f-b8aa-dbb52855725b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "369"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "03d60243-323f-4c6a-af58-658f41509499",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "366 \n"
     ]
    }
   ],
   "source": [
    "pncA_seq = h37rv_seq[h37rv_summary.loc[\"pncA\", \"chrom_start\"]-1:h37rv_summary.loc[\"pncA\", \"chrom_stop\"]]\n",
    "\n",
    "for i in range(len(l3_pncA_seq)):\n",
    "    if pncA_seq[i] != l3_pncA_seq[i]:\n",
    "        print(i, end=\" \")\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc81a690-fdba-4ab5-982b-c606c765ce82",
   "metadata": {},
   "source": [
    "## Create list of mutate_ix for each lineage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1ae2e4ce-d7ea-4283-b7b9-aad391d72527",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>mut_base</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGAAGTGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGAATAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGTAGAACACCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>A</td>\n",
       "      <td>GGAAGTCCTTGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix mut_base             kmer\n",
       "0   2288777        G      CCAGCACCCTG\n",
       "1   2289071        G    GAGAAGTGGTCAC\n",
       "2   2289049        G      GAGGAATAGTC\n",
       "3   2288961        C  TTGTAGAACACCGCC\n",
       "4   2289096        A    GGAAGTCCTTGGT"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DR_mut_kmers_df = pd.read_csv(\"./DR_mut_h37rv_kmers.csv\")\n",
    "DR_mut_kmers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "282cfa2b-9b4d-4d01-978c-df8091eeed3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l1_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L1_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L1_seq += line.strip()\n",
    "        \n",
    "with open(l2_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L2_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L2_seq += line.strip()\n",
    "        \n",
    "with open(l3_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L3_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L3_seq += line.strip()\n",
    "\n",
    "with open(l4_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L4_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L4_seq += line.strip()\n",
    "        \n",
    "L_seqs = {1: L1_seq, 2: L2_seq, 3: L3_seq, 4: L4_seq}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "90be7b3e-7b44-4ea5-a0ca-bf49dc1f7528",
   "metadata": {},
   "outputs": [],
   "source": [
    "problematic_ix = {1: [7581], 2: [7581], 3: [7581, 2289049], 4: [7581]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d9c27ae2-d58f-48b1-b530-2bc4caf814a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "problematic_ix_gene_map = {7581: \"gyrA\", 2289049: \"pncA\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "478e5ab1-aa21-44de-93b5-b6e4a8920cc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>L1_ix</th>\n",
       "      <th>L2_ix</th>\n",
       "      <th>L3_ix</th>\n",
       "      <th>L4_ix</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>mut_base</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2288777</th>\n",
       "      <td>2310818</td>\n",
       "      <td>2288812</td>\n",
       "      <td>2298581</td>\n",
       "      <td>2301252</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2289071</th>\n",
       "      <td>2311112</td>\n",
       "      <td>2289106</td>\n",
       "      <td>2298875</td>\n",
       "      <td>2301546</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2289049</th>\n",
       "      <td>2311090</td>\n",
       "      <td>2289084</td>\n",
       "      <td>2298853</td>\n",
       "      <td>2301524</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2288961</th>\n",
       "      <td>2311002</td>\n",
       "      <td>2288996</td>\n",
       "      <td>2298765</td>\n",
       "      <td>2301436</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2289096</th>\n",
       "      <td>2311137</td>\n",
       "      <td>2289131</td>\n",
       "      <td>2298900</td>\n",
       "      <td>2301571</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2288919</th>\n",
       "      <td>2310960</td>\n",
       "      <td>2288954</td>\n",
       "      <td>2298723</td>\n",
       "      <td>2301394</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7569</th>\n",
       "      <td>7569</td>\n",
       "      <td>8927</td>\n",
       "      <td>7569</td>\n",
       "      <td>7569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2288816</th>\n",
       "      <td>2310857</td>\n",
       "      <td>2288851</td>\n",
       "      <td>2298620</td>\n",
       "      <td>2301291</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4247468</th>\n",
       "      <td>4257279</td>\n",
       "      <td>4257016</td>\n",
       "      <td>4257985</td>\n",
       "      <td>4252610</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4407795</th>\n",
       "      <td>4417667</td>\n",
       "      <td>4417432</td>\n",
       "      <td>4419779</td>\n",
       "      <td>4412978</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761138</th>\n",
       "      <td>765845</td>\n",
       "      <td>760123</td>\n",
       "      <td>768182</td>\n",
       "      <td>762800</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2288838</th>\n",
       "      <td>2310879</td>\n",
       "      <td>2288873</td>\n",
       "      <td>2298642</td>\n",
       "      <td>2301313</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4327266</th>\n",
       "      <td>4337064</td>\n",
       "      <td>4336815</td>\n",
       "      <td>4339142</td>\n",
       "      <td>4332410</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1473245</th>\n",
       "      <td>1479933</td>\n",
       "      <td>1473300</td>\n",
       "      <td>1478127</td>\n",
       "      <td>1473339</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7581</th>\n",
       "      <td>7581</td>\n",
       "      <td>8939</td>\n",
       "      <td>7581</td>\n",
       "      <td>7581</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2288972</th>\n",
       "      <td>2311013</td>\n",
       "      <td>2289007</td>\n",
       "      <td>2298776</td>\n",
       "      <td>2301447</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2288886</th>\n",
       "      <td>2310927</td>\n",
       "      <td>2288921</td>\n",
       "      <td>2298690</td>\n",
       "      <td>2301361</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2289104</th>\n",
       "      <td>2311145</td>\n",
       "      <td>2289139</td>\n",
       "      <td>2298908</td>\n",
       "      <td>2301579</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761136</th>\n",
       "      <td>765843</td>\n",
       "      <td>760121</td>\n",
       "      <td>768180</td>\n",
       "      <td>762798</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2289102</th>\n",
       "      <td>2311143</td>\n",
       "      <td>2289137</td>\n",
       "      <td>2298906</td>\n",
       "      <td>2301577</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            L1_ix    L2_ix    L3_ix    L4_ix ref_base mut_base\n",
       "h37rv_ix                                                      \n",
       "2288777   2310818  2288812  2298581  2301252        A        G\n",
       "2289071   2311112  2289106  2298875  2301546        T        G\n",
       "2289049   2311090  2289084  2298853  2301524        A        G\n",
       "2288961   2311002  2288996  2298765  2301436        A        C\n",
       "2289096   2311137  2289131  2298900  2301571        C        A\n",
       "2288919   2310960  2288954  2298723  2301394        C        A\n",
       "7569         7569     8927     7569     7569        C        T\n",
       "2288816   2310857  2288851  2298620  2301291        G        T\n",
       "4247468   4257279  4257016  4257985  4252610        A        G\n",
       "4407795   4417667  4417432  4419779  4412978        G        A\n",
       "761138     765845   760123   768182   762800        C        G\n",
       "2288838   2310879  2288873  2298642  2301313        T        C\n",
       "4327266   4337064  4336815  4339142  4332410        C        A\n",
       "1473245   1479933  1473300  1478127  1473339        A        G\n",
       "7581         7581     8939     7581     7581        A        C\n",
       "2288972   2311013  2289007  2298776  2301447        A        C\n",
       "2288886   2310927  2288921  2298690  2301361        A        C\n",
       "2289104   2311145  2289139  2298908  2301579        G        T\n",
       "761136     765843   760121   768180   762798        C        T\n",
       "2289102   2311143  2289137  2298906  2301577        T        G"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c = pd.DataFrame(columns=[\"h37rv_ix\", \"L1_ix\", \"L2_ix\", \"L3_ix\", \"L4_ix\", \"ref_base\", \"mut_base\"]).set_index(\"h37rv_ix\")\n",
    "\n",
    "for i in DR_mut_kmers_df.index:\n",
    "    \n",
    "    h37rv_ix = DR_mut_kmers_df.loc[i, \"h37rv_ix\"]\n",
    "    h37rv_base = h37rv_seq[h37rv_ix]\n",
    "    mut_base = DR_mut_kmers_df.loc[i, \"mut_base\"]\n",
    "    \n",
    "    kmer = DR_mut_kmers_df.loc[i, \"kmer\"]\n",
    "    kmer_length = len(kmer)\n",
    "    \n",
    "    L_ix = {}\n",
    "    for lin_i in range(1, 5):\n",
    "        \n",
    "        if h37rv_ix not in problematic_ix[lin_i]:\n",
    "            match_start = L_seqs[lin_i].find(kmer)    \n",
    "            L_ix[lin_i] = match_start + (kmer_length-1)//2\n",
    "\n",
    "        else:\n",
    "            gene = problematic_ix_gene_map[h37rv_ix]\n",
    "            gene_h37rv_start = h37rv_summary.loc[gene, \"chrom_start\"]\n",
    "            gene_ix = h37rv_ix - gene_h37rv_start + 1\n",
    "            \n",
    "            blast_df = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/lineage123/blast/lineage{}.{}.blastn.txt\".format(lin_i, gene),\n",
    "                                sep=\"\\t\",\n",
    "                                names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\", \"qseq\", \"nident\"])\n",
    "            \n",
    "            L_start_ix = blast_df.loc[0, \"qstart\"] #1-indexed\n",
    "            L_ix[lin_i] = L_start_ix + gene_ix - 1\n",
    "    \n",
    "    \n",
    "        assert(h37rv_base == L_seqs[lin_i][L_ix[lin_i]])\n",
    "        \n",
    "    L1234_mut_ix_df.loc[h37rv_ix] = [L_ix[1], L_ix[2], L_ix[3], L_ix[4], h37rv_base, mut_base]\n",
    "    \n",
    "L1234_mut_ix_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7c161fc7-bc63-41c2-9edc-92cebfe299b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1234_mut_ix_df.to_csv(\"./DR_mut_lineage_ix.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ea3a12-8f3f-4e5e-b1a1-e9e0ebaed71c",
   "metadata": {},
   "source": [
    "# Modification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "68d141ad-7c0e-4650-a4c6-e5abf187399b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>L1_ix</th>\n",
       "      <th>L2_ix</th>\n",
       "      <th>L3_ix</th>\n",
       "      <th>L4_ix</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>mut_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>2310818</td>\n",
       "      <td>2288812</td>\n",
       "      <td>2298581</td>\n",
       "      <td>2301252</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>2311112</td>\n",
       "      <td>2289106</td>\n",
       "      <td>2298875</td>\n",
       "      <td>2301546</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>2311090</td>\n",
       "      <td>2289084</td>\n",
       "      <td>2298853</td>\n",
       "      <td>2301524</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>2311002</td>\n",
       "      <td>2288996</td>\n",
       "      <td>2298765</td>\n",
       "      <td>2301436</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>2311137</td>\n",
       "      <td>2289131</td>\n",
       "      <td>2298900</td>\n",
       "      <td>2301571</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix    L1_ix    L2_ix    L3_ix    L4_ix ref_base mut_base\n",
       "0   2288777  2310818  2288812  2298581  2301252        A        G\n",
       "1   2289071  2311112  2289106  2298875  2301546        T        G\n",
       "2   2289049  2311090  2289084  2298853  2301524        A        G\n",
       "3   2288961  2311002  2288996  2298765  2301436        A        C\n",
       "4   2289096  2311137  2289131  2298900  2301571        C        A"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L123_mut_ix_df = pd.read_csv(\"./DR_mut_lineage_ix.csv\")\n",
    "L123_mut_ix_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f1a00420-89d8-4749-9fbd-d06abceb8907",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l1_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L1_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L1_seq += line.strip()\n",
    "        \n",
    "with open(l2_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L2_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L2_seq += line.strip()\n",
    "        \n",
    "with open(l3_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L3_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L3_seq += line.strip()\n",
    "        \n",
    "L1_seq_array = [x for x in L1_seq]\n",
    "L2_seq_array = [x for x in L2_seq]\n",
    "L3_seq_array = [x for x in L3_seq]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f4a16d55-15bf-4710-aacb-52b55b81150a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in L123_mut_ix_df.index:\n",
    "    \n",
    "    L1_ix = L123_mut_ix_df.loc[i, \"L1_ix\"]\n",
    "    L2_ix = L123_mut_ix_df.loc[i, \"L2_ix\"]\n",
    "    L3_ix = L123_mut_ix_df.loc[i, \"L3_ix\"]\n",
    "    \n",
    "    ref_base, mut_base = L123_mut_ix_df.loc[i, \"ref_base\"], L123_mut_ix_df.loc[i, \"mut_base\"]\n",
    "    \n",
    "    assert(L1_seq_array[L1_ix] == ref_base)\n",
    "    L1_seq_array[L1_ix] = mut_base\n",
    "    \n",
    "    assert(L2_seq_array[L2_ix] == ref_base)\n",
    "    L2_seq_array[L2_ix] = mut_base\n",
    "    \n",
    "    assert(L3_seq_array[L3_ix] == ref_base)\n",
    "    L3_seq_array[L3_ix] = mut_base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8be04afb-9c89-48a0-b97c-779143c70ba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"./fastas_final\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "24b7ccc1-d134-409c-9eb1-6b52b8ec3d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1_seq_main = \"\".join(L1_seq_array)\n",
    "with open(\"{}/L1_mutant/L1_mutant.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L1-mutant\\n\")\n",
    "    n_chunks = len(L1_seq_main)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L1_seq_main[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L1_seq_main[(i+1)*80:] + \"\\n\")\n",
    "    \n",
    "L2_seq_main = \"\".join(L2_seq_array)\n",
    "with open(\"{}/L2_mutant/L2_mutant.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L2-mutant\\n\")\n",
    "    n_chunks = len(L2_seq_main)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L2_seq_main[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L2_seq_main[(i+1)*80:] + \"\\n\")\n",
    "\n",
    "L3_seq_main = \"\".join(L3_seq_array)\n",
    "with open(\"{}/L3_mutant/L3_mutant.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L3-mutant\\n\")\n",
    "    n_chunks = len(L3_seq_main)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L3_seq_main[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L3_seq_main[(i+1)*80:] + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "74034e1f-1dc3-4945-b8be-b01315676ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"{}/L1_mutant/L1.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L1\\n\")\n",
    "    n_chunks = len(L1_seq)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L1_seq[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L1_seq[(i+1)*80:] + \"\\n\")\n",
    "    \n",
    "with open(\"{}/L2_mutant/L2.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L2\\n\")\n",
    "    n_chunks = len(L2_seq)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L2_seq[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L2_seq[(i+1)*80:] + \"\\n\")\n",
    "\n",
    "with open(\"{}/L3_mutant/L3.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L3\\n\")\n",
    "    n_chunks = len(L3_seq)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L3_seq[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L3_seq[(i+1)*80:] + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b6a8d698-eb96-4841-a6d6-3d66e77a57ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>L1_ix</th>\n",
       "      <th>L2_ix</th>\n",
       "      <th>L3_ix</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>mut_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>2310818</td>\n",
       "      <td>2288812</td>\n",
       "      <td>2298581</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>2311112</td>\n",
       "      <td>2289106</td>\n",
       "      <td>2298875</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>2311090</td>\n",
       "      <td>2289084</td>\n",
       "      <td>2298853</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>2311002</td>\n",
       "      <td>2288996</td>\n",
       "      <td>2298765</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>2311137</td>\n",
       "      <td>2289131</td>\n",
       "      <td>2298900</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix    L1_ix    L2_ix    L3_ix ref_base mut_base\n",
       "0   2288777  2310818  2288812  2298581        A        G\n",
       "1   2289071  2311112  2289106  2298875        T        G\n",
       "2   2289049  2311090  2289084  2298853        A        G\n",
       "3   2288961  2311002  2288996  2298765        A        C\n",
       "4   2289096  2311137  2289131  2298900        C        A"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L123_mut_ix_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "14955f00-8e86-4e69-b89f-d0d7211bec64",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1_df = pd.DataFrame(columns=[\"POS\", \"REF\", \"ALT\", \"TYPE\"])\n",
    "L1_df[\"POS\"] = L123_mut_ix_df.L1_ix+1\n",
    "L1_df[\"REF\"] = L123_mut_ix_df.ref_base\n",
    "L1_df[\"ALT\"] = L123_mut_ix_df.mut_base\n",
    "L1_df[\"TYPE\"] = [\"DR\"]*L123_mut_ix_df.shape[0]\n",
    "\n",
    "L2_df = pd.DataFrame(columns=[\"POS\", \"REF\", \"ALT\", \"TYPE\"])\n",
    "L2_df[\"POS\"] = L123_mut_ix_df.L2_ix+1\n",
    "L2_df[\"REF\"] = L123_mut_ix_df.ref_base\n",
    "L2_df[\"ALT\"] = L123_mut_ix_df.mut_base\n",
    "L2_df[\"TYPE\"] = [\"DR\"]*L123_mut_ix_df.shape[0]\n",
    "\n",
    "L3_df = pd.DataFrame(columns=[\"POS\", \"REF\", \"ALT\", \"TYPE\"])\n",
    "L3_df[\"POS\"] = L123_mut_ix_df.L3_ix+1\n",
    "L3_df[\"REF\"] = L123_mut_ix_df.ref_base\n",
    "L3_df[\"ALT\"] = L123_mut_ix_df.mut_base\n",
    "L3_df[\"TYPE\"] = [\"DR\"]*L123_mut_ix_df.shape[0]\n",
    "\n",
    "L1_df.to_csv(\"./mutant_summaries_final/L1_mutant.csv\", index=False)\n",
    "L2_df.to_csv(\"./mutant_summaries_final/L2_mutant.csv\", index=False)\n",
    "L3_df.to_csv(\"./mutant_summaries_final/L3_mutant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "977bfbf0-18fd-4bd0-8e33-5323df2c1b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h37rv_ix</th>\n",
       "      <th>L1_ix</th>\n",
       "      <th>L2_ix</th>\n",
       "      <th>L3_ix</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>mut_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2288777</td>\n",
       "      <td>2310818</td>\n",
       "      <td>2288812</td>\n",
       "      <td>2298581</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2289071</td>\n",
       "      <td>2311112</td>\n",
       "      <td>2289106</td>\n",
       "      <td>2298875</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2289049</td>\n",
       "      <td>2311090</td>\n",
       "      <td>2289084</td>\n",
       "      <td>2298853</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2288961</td>\n",
       "      <td>2311002</td>\n",
       "      <td>2288996</td>\n",
       "      <td>2298765</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2289096</td>\n",
       "      <td>2311137</td>\n",
       "      <td>2289131</td>\n",
       "      <td>2298900</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   h37rv_ix    L1_ix    L2_ix    L3_ix ref_base mut_base\n",
       "0   2288777  2310818  2288812  2298581        A        G\n",
       "1   2289071  2311112  2289106  2298875        T        G\n",
       "2   2289049  2311090  2289084  2298853        A        G\n",
       "3   2288961  2311002  2288996  2298765        A        C\n",
       "4   2289096  2311137  2289131  2298900        C        A"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L123_mut_ix_df = pd.read_csv(\"./DR_mut_lineage_ix.csv\")\n",
    "L123_mut_ix_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5baf023f-0607-4691-a2bb-a710e3c2827d",
   "metadata": {},
   "outputs": [],
   "source": [
    "L123_df = pd.DataFrame(columns=[\"POS\", \"REF\", \"ALT\", \"TYPE\"])\n",
    "L123_df[\"POS\"] = L123_mut_ix_df.h37rv_ix+1\n",
    "L123_df[\"REF\"] = L123_mut_ix_df.ref_base\n",
    "L123_df[\"ALT\"] = L123_mut_ix_df.mut_base\n",
    "L123_df[\"TYPE\"] = [\"DR\"]*L123_mut_ix_df.shape[0]\n",
    "\n",
    "L123_df.to_csv(\"./mutant_summaries_final/L123_mutant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57223bf9-85f5-47f8-9510-876ca1738342",
   "metadata": {},
   "source": [
    "## L4 addition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "751745ca-9288-4a11-82ab-03e2f9b09c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(l4_fa_path, \"r\") as in_f:\n",
    "    header = in_f.readline()\n",
    "    \n",
    "    line = in_f.readline()\n",
    "    L4_seq = line.strip()\n",
    "    while line:\n",
    "        line = in_f.readline()\n",
    "        L4_seq += line.strip()\n",
    "        \n",
    "L4_seq_array = [x for x in L4_seq]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e29c60a5-6942-49d3-a659-0646243a9cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in L123_mut_ix_df.index:\n",
    "    \n",
    "    L4_ix = L123_mut_ix_df.loc[i, \"L4_ix\"]\n",
    "    \n",
    "    ref_base, mut_base = L123_mut_ix_df.loc[i, \"ref_base\"], L123_mut_ix_df.loc[i, \"mut_base\"]\n",
    "    \n",
    "    assert(L4_seq_array[L4_ix] == ref_base)\n",
    "    L4_seq_array[L4_ix] = mut_base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "bdfe52b2-4532-48c9-a20b-423e783361d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"./fastas_final\"\n",
    "\n",
    "L4_seq_main = \"\".join(L4_seq_array)\n",
    "with open(\"{}/L4_mutant/L4_mutant.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L4-mutant\\n\")\n",
    "    n_chunks = len(L4_seq_main)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L4_seq_main[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L4_seq_main[(i+1)*80:] + \"\\n\")\n",
    "    \n",
    "with open(\"{}/L4_mutant/L4.fasta\".format(out_dir), \"w\") as out_f:\n",
    "    out_f.write(\">L4\\n\")\n",
    "    n_chunks = len(L4_seq)//80\n",
    "    for i in range(n_chunks):\n",
    "        out_f.write(L4_seq[i*80:i*80+80] + \"\\n\")\n",
    "    out_f.write(L4_seq[(i+1)*80:] + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "671deedb-c255-4068-8cba-4c7ef0919b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "L4_df = pd.DataFrame(columns=[\"POS\", \"REF\", \"ALT\", \"TYPE\"])\n",
    "L4_df[\"POS\"] = L123_mut_ix_df.L4_ix+1\n",
    "L4_df[\"REF\"] = L123_mut_ix_df.ref_base\n",
    "L4_df[\"ALT\"] = L123_mut_ix_df.mut_base\n",
    "L4_df[\"TYPE\"] = [\"DR\"]*L123_mut_ix_df.shape[0]\n",
    "\n",
    "L4_df.to_csv(\"./mutant_summaries_final/L4_mutant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "337a8a3e-b48f-4007-b9a8-bf759feaaf05",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
