{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0628a7b-5202-420d-8edb-5b0369a326b9",
   "metadata": {},
   "source": [
    "# FILTER FP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ed9c86f7-66e0-4db7-8f57-bd1b90f2f727",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import statsmodels.stats.multicomp as mc\n",
    "from scipy import stats\n",
    "import pickle\n",
    "from scipy.stats import mannwhitneyu as mannwu\n",
    "from matplotlib.ticker import MultipleLocator\n",
    "import os\n",
    "\n",
    "import sys\n",
    "sys.path.append(\"./scripts/modules\")\n",
    "\n",
    "from benchmarking_definitions import *\n",
    "from plotting import *\n",
    "from regions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "43b8ce65-115b-4a8f-98ce-718a55afb963",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2500, 1000)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tagsRv = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/FINAL_tag_list.csv\", names=[\"tag\"])[\"tag\"].values\n",
    "tagsL1234 = pd.read_csv(\"/home/sm624/projects/mixed_calls/benchmarking/comprehensive/FINAL_tag_list_L1234.csv\", names=[\"tag\"]).tag.values\n",
    "\n",
    "len(tagsRv), len(tagsL1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e6b8bfbf-6d88-46b0-afcc-9127bf93468c",
   "metadata": {},
   "outputs": [],
   "source": [
    "AF_min = 0.05\n",
    "AF_max = 0.95\n",
    "MQ_thresh = 40\n",
    "num_support_each_direction = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cdc4925a-a07f-437a-adaf-3586ea1a83e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "h37Rv_path = \"/n/data1/hms/dbmi/farhat/Sanjana/H37Rv\"\n",
    "h37Rv_regions = pd.read_csv(os.path.join(h37Rv_path, \"mycobrowser_h37rv_v4.csv\"))\n",
    "\n",
    "# remove rRNAs, which are highly conserved. rrs, rrl, and rrf\n",
    "rRNA_pos = []\n",
    "\n",
    "for i, row in h37Rv_regions.query(\"Functional_Category=='stable RNAs' & Feature=='rRNA'\").iterrows():\n",
    "    rRNA_pos += list(np.arange(row['Start'], row['Stop'] + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "090b962c-1d03-4657-a3e5-9b3861b2c8ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Refseq_ID</th>\n",
       "      <th>Mycobrowser_Version</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Name</th>\n",
       "      <th>...</th>\n",
       "      <th>SWISS-MODEL</th>\n",
       "      <th>Orthologues M. leprae</th>\n",
       "      <th>Orthologues M. marinum</th>\n",
       "      <th>Orthologues M. smegmatis</th>\n",
       "      <th>Orthologues M. bovis</th>\n",
       "      <th>Orthologues M. lepromatosis</th>\n",
       "      <th>Orthologues M. tuberculosis</th>\n",
       "      <th>Orthologues M. abscessus</th>\n",
       "      <th>Orthologues M. haemophilum</th>\n",
       "      <th>Orthologues M. orygis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1417</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>Mycobrowser_v4</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1471846</td>\n",
       "      <td>1473382</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MTB000019</td>\n",
       "      <td>rrs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1419</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>Mycobrowser_v4</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1473658</td>\n",
       "      <td>1476795</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MTB000020</td>\n",
       "      <td>rrl</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1420</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>Mycobrowser_v4</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1476899</td>\n",
       "      <td>1477013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MTB000021</td>\n",
       "      <td>rrf</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Refseq_ID Mycobrowser_Version Feature    Start     Stop Score Strand  \\\n",
       "1417  NC_000962.3      Mycobrowser_v4    rRNA  1471846  1473382     .      +   \n",
       "1419  NC_000962.3      Mycobrowser_v4    rRNA  1473658  1476795     .      +   \n",
       "1420  NC_000962.3      Mycobrowser_v4    rRNA  1476899  1477013     .      +   \n",
       "\n",
       "      Frame      Locus Name  ... SWISS-MODEL Orthologues M. leprae  \\\n",
       "1417    0.0  MTB000019  rrs  ...         NaN                   NaN   \n",
       "1419    0.0  MTB000020  rrl  ...         NaN                   NaN   \n",
       "1420    0.0  MTB000021  rrf  ...         NaN                   NaN   \n",
       "\n",
       "     Orthologues M. marinum Orthologues M. smegmatis Orthologues M. bovis  \\\n",
       "1417                    NaN                      NaN                  NaN   \n",
       "1419                    NaN                      NaN                  NaN   \n",
       "1420                    NaN                      NaN                  NaN   \n",
       "\n",
       "     Orthologues M. lepromatosis Orthologues M. tuberculosis  \\\n",
       "1417                         NaN                         NaN   \n",
       "1419                         NaN                         NaN   \n",
       "1420                         NaN                         NaN   \n",
       "\n",
       "     Orthologues M. abscessus Orthologues M. haemophilum Orthologues M. orygis  \n",
       "1417                      NaN                        NaN                   NaN  \n",
       "1419                      NaN                        NaN                   NaN  \n",
       "1420                      NaN                        NaN                   NaN  \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h37Rv_regions.query(\"Functional_Category=='stable RNAs' & Feature=='rRNA'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "78995d53-b2fe-46f5-8e33-ec0474d9b734",
   "metadata": {},
   "outputs": [],
   "source": [
    "excl_cols = [\"EXCL_DP\", \"EXCL_MQ\", \"EXCL_SAO\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "766ce9f4-023c-4296-a7e5-3fa6d1938633",
   "metadata": {},
   "source": [
    "# ISS L1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3d9031fb-44e7-4594-bc90-d86b7279542e",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_L1234\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "038e8313-e1cf-4a54-bfef-ea2821c14ab0",
   "metadata": {},
   "source": [
    "## Filter variant summary file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "409c2f1c-3213-4513-9bb4-faac7e8fa1e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "source_dir = f\"{parent_dir}/position_metrics\"\n",
    "output_dir = f\"{parent_dir}/SNPs\"\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    position_metric_df = pd.read_csv(f\"{source_dir}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "    position_metric_df = position_metric_df[position_metric_df.L_GT == 0] ## L1234-specific\n",
    "\n",
    "    # filter for only the variants found by FreeBayes\n",
    "    position_metric_df = position_metric_df[position_metric_df.FB_found == 1]\n",
    "\n",
    "    # extract necessary columns\n",
    "    _position_metric_df = position_metric_df[[\"POS\", \"REF\", \"ALT\", \"GT\", \"REGION\", \"FB_AF\", \"COV_RATIO\", \"CLIPPED_BASES_RATIO\", \"DISCORDANT_READS_RATIO\",\n",
    "                                              \"MEAN_ALT_BQ\", \"FB_STRAND_BIAS\", \"FB_AO\", \"FB_AO_F\", \"FB_AO_R\", \"FB_DP\", \"FB_MQ\", \"FB_QUAL\",\n",
    "                                              \"FB_ORIGIN\", \"FB_COMPLEX_TYPE\", \"FB_MULTIPLE_HAPS\", \"FB_MULTIALLELIC\", \"FB_REF_GL\", \"FB_ALT_GL\"]]\n",
    "\n",
    "    # filter SNPs\n",
    "    SNP_df = _position_metric_df[(_position_metric_df.REF.str.len() == 1) & (_position_metric_df.ALT.str.len() == 1)].copy(deep=True)\n",
    "\n",
    "    # rename columns for consistency with variables used by filtering model\n",
    "    SNP_df = SNP_df.rename({\"MEAN_ALT_BQ\":\"Mean_BQ_ALT_allele\",\n",
    "                            \"FB_STRAND_BIAS\": \"SAF_prop_deviation_from_half\"}, axis=1)\n",
    "\n",
    "    # already split MNPs\n",
    "\n",
    "    SNP_df.to_csv(f\"{output_dir}/{tag}.candidate_SNPs.csv\", index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc0810e8-8fbf-4af6-ac14-7e0f7ef71367",
   "metadata": {},
   "source": [
    "## Make tag ID file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "57153667-4a72-4c93-883d-f315874ad151",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "tag_id_df = pd.DataFrame(columns=[\"tag\"])\n",
    "tag_id_df.loc[0, :] = 1\n",
    "tag_id_df.loc[0, \"tag\"] = \"dummy\"\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    lowAF_SNP_file = f\"{output_dir}/{tag}.candidate_SNPs.csv\"\n",
    "\n",
    "    lowAF_SNP_df = pd.read_csv(lowAF_SNP_file)\n",
    "\n",
    "    add_df = pd.DataFrame(columns=[\"tag\"])\n",
    "    add_df[\"tag\"] = [tag]*lowAF_SNP_df.shape[0]\n",
    "\n",
    "    tag_id_df = pd.concat([tag_id_df, add_df])\n",
    "    tag_id_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "tag_id_df.drop(0, inplace=True)\n",
    "tag_id_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b87fd4f8-8296-4110-8b41-d50c76420d9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1057642, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag\n",
       "0  sim1_L1_mutant_50_0.01\n",
       "1  sim1_L1_mutant_50_0.01"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(tag_id_df.shape)\n",
    "tag_id_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "886a0aa1-50c8-4d04-bc7a-62c6c2a9ac4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tag_id_df.to_csv(\"/n/scratch/users/s/sm624/L1234_benchmarking_tag_ids.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b9ec34b-4f14-4f5e-9f9d-50829e77b54d",
   "metadata": {},
   "source": [
    "## Create final file of SNPs with predictions and add exclusions\n",
    "\n",
    "1. Label low-frequency SNP (AF ≤ 95%) and fixed SNP (AF > 95%).\n",
    "2. Label fixed exclusion filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "894d63e2-c92c-402c-9a95-2aa6cb00efe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_L1234\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5705bcd6-e9b9-47c5-b7eb-e9b6b45fd799",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1057642, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GT</th>\n",
       "      <th>REGION</th>\n",
       "      <th>FB_AF</th>\n",
       "      <th>COV_RATIO</th>\n",
       "      <th>CLIPPED_BASES_RATIO</th>\n",
       "      <th>DISCORDANT_READS_RATIO</th>\n",
       "      <th>...</th>\n",
       "      <th>FB_MQ</th>\n",
       "      <th>FB_QUAL</th>\n",
       "      <th>FB_ORIGIN</th>\n",
       "      <th>FB_COMPLEX_TYPE</th>\n",
       "      <th>FB_MULTIPLE_HAPS</th>\n",
       "      <th>FB_MULTIALLELIC</th>\n",
       "      <th>FB_REF_GL</th>\n",
       "      <th>FB_ALT_GL</th>\n",
       "      <th>predicted</th>\n",
       "      <th>pred_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>21455</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>other</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.956149</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>60.0</td>\n",
       "      <td>3.887680e-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-157.967</td>\n",
       "      <td>0.977903</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>24721</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>LM</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.706861</td>\n",
       "      <td>0.264706</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>24721.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.005015</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag    POS REF ALT  GT REGION     FB_AF  COV_RATIO  \\\n",
       "0  sim1_L1_mutant_50_0.01  21455   C   T   0  other  0.038462   0.956149   \n",
       "1  sim1_L1_mutant_50_0.01  24721   G   C   0     LM  0.062500   0.706861   \n",
       "\n",
       "   CLIPPED_BASES_RATIO  DISCORDANT_READS_RATIO  ...  FB_MQ       FB_QUAL  \\\n",
       "0             0.000000                     0.0  ...   60.0  3.887680e-15   \n",
       "1             0.264706                     0.0  ...   60.0  0.000000e+00   \n",
       "\n",
       "   FB_ORIGIN  FB_COMPLEX_TYPE  FB_MULTIPLE_HAPS  FB_MULTIALLELIC  FB_REF_GL  \\\n",
       "0        NaN                0                 0                0        0.0   \n",
       "1    24721.0                1                 0                0        NaN   \n",
       "\n",
       "   FB_ALT_GL  predicted  pred_class  \n",
       "0   -157.967   0.977903         1.0  \n",
       "1        NaN   0.005015         0.0  \n",
       "\n",
       "[2 rows x 26 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicted_MV_SNPs_file = f\"{parent_dir}/variant_SNP_predictions.csv\"\n",
    "predicted_MV_SNPs_df = pd.read_csv(predicted_MV_SNPs_file, low_memory=False)\n",
    "print(predicted_MV_SNPs_df.shape)\n",
    "predicted_MV_SNPs_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "938f319d-f403-4890-85b4-daf9ff876610",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "source_dir = f\"{parent_dir}/position_metrics\"\n",
    "output_dir = f\"{parent_dir}/SNPs\"\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    SNP_df = predicted_MV_SNPs_df[predicted_MV_SNPs_df.tag == tag]\n",
    "    SNP_df = SNP_df.drop(\"tag\", axis=1)\n",
    "\n",
    "    # label with exclusion filters\n",
    "    SNP_df[\"EXCL_DP\"] = (SNP_df.FB_DP < 5).astype(int)\n",
    "    SNP_df[\"EXCL_MQ\"] = (SNP_df.FB_MQ < 40).astype(int)\n",
    "    SNP_df[\"EXCL_SAO\"] = ((SNP_df.FB_AO_F < 2) | (SNP_df.FB_AO_R < 2)).astype(int)\n",
    "    \n",
    "    SNP_df.to_csv(f\"{output_dir}/{tag}.SNPs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95978ce2-e131-4940-ac5f-66649b8a2a85",
   "metadata": {},
   "source": [
    "# ISS H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83bff60a-25d0-4908-805d-48fbfc29a394",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_H37Rv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a821a0b-cd9b-4e0d-ba4c-636910a293ae",
   "metadata": {},
   "source": [
    "## Filter variant summary file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cc2bddfe-2b0e-4d09-8437-568941d8af85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "source_dir = f\"{parent_dir}/position_metrics\"\n",
    "output_dir = f\"{parent_dir}/SNPs\"\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    position_metric_df = pd.read_csv(f\"{source_dir}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "\n",
    "    # filter for only the variants found by FreeBayes\n",
    "    position_metric_df = position_metric_df[position_metric_df.FB_found == 1]\n",
    "\n",
    "    # extract necessary columns\n",
    "    _position_metric_df = position_metric_df[[\"POS\", \"REF\", \"ALT\", \"GT\", \"REGION\", \"FB_AF\", \"COV_RATIO\", \"CLIPPED_BASES_RATIO\", \"DISCORDANT_READS_RATIO\",\n",
    "                                              \"MEAN_ALT_BQ\", \"FB_STRAND_BIAS\", \"FB_AO\", \"FB_AO_F\", \"FB_AO_R\", \"FB_DP\", \"FB_MQ\", \"FB_QUAL\",\n",
    "                                              \"FB_ORIGIN\", \"FB_COMPLEX_TYPE\", \"FB_MULTIPLE_HAPS\", \"FB_MULTIALLELIC\", \"FB_REF_GL\", \"FB_ALT_GL\"]]\n",
    "\n",
    "    # filter SNPs\n",
    "    SNP_df = _position_metric_df[(_position_metric_df.REF.str.len() == 1) & (_position_metric_df.ALT.str.len() == 1)].copy(deep=True)\n",
    "\n",
    "    # rename columns for consistency with variables used by filtering model\n",
    "    SNP_df = SNP_df.rename({\"MEAN_ALT_BQ\":\"Mean_BQ_ALT_allele\",\n",
    "                            \"FB_STRAND_BIAS\": \"SAF_prop_deviation_from_half\"}, axis=1)\n",
    "\n",
    "    # already split MNPs\n",
    "\n",
    "    SNP_df.to_csv(f\"{output_dir}/{tag}.candidate_SNPs.csv\", index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ba6bb0-462f-4d65-bd8d-9b18634006eb",
   "metadata": {},
   "source": [
    "## Make tag ID file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c6f864c8-b56b-4f71-8501-ced83b4486de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "tag_id_df = pd.DataFrame(columns=[\"tag\"])\n",
    "tag_id_df.loc[0, :] = 1\n",
    "tag_id_df.loc[0, \"tag\"] = \"dummy\"\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    lowAF_SNP_file = f\"{output_dir}/{tag}.candidate_SNPs.csv\"\n",
    "\n",
    "    lowAF_SNP_df = pd.read_csv(lowAF_SNP_file)\n",
    "\n",
    "    add_df = pd.DataFrame(columns=[\"tag\"])\n",
    "    add_df[\"tag\"] = [tag]*lowAF_SNP_df.shape[0]\n",
    "\n",
    "    tag_id_df = pd.concat([tag_id_df, add_df])\n",
    "    tag_id_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "tag_id_df.drop(0, inplace=True)\n",
    "tag_id_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a1ebb41d-0017-4621-a90c-7cfce25b4634",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(496843, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag\n",
       "0  sim1_h37rv_mutant_1_50_0.01\n",
       "1  sim1_h37rv_mutant_1_50_0.01"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(tag_id_df.shape)\n",
    "tag_id_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "04be1ddf-2458-45e9-99b9-dadc0ce74635",
   "metadata": {},
   "outputs": [],
   "source": [
    "tag_id_df.to_csv(\"/n/scratch/users/s/sm624/H37Rv_benchmarking_tag_ids.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59faaf29-731c-46c3-a4b2-1b31cd784b49",
   "metadata": {},
   "source": [
    "## Create final file of SNPs with predictions and add exclusions\n",
    "\n",
    "1. Label low-frequency SNP (AF ≤ 95%) and fixed SNP (AF > 95%).\n",
    "2. Label fixed exclusion filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6b178309-86fd-468a-b007-c57242fe9de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_H37Rv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d5070daa-b955-446b-ad77-c2945b914c2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(496843, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GT</th>\n",
       "      <th>REGION</th>\n",
       "      <th>FB_AF</th>\n",
       "      <th>COV_RATIO</th>\n",
       "      <th>CLIPPED_BASES_RATIO</th>\n",
       "      <th>DISCORDANT_READS_RATIO</th>\n",
       "      <th>...</th>\n",
       "      <th>FB_MQ</th>\n",
       "      <th>FB_QUAL</th>\n",
       "      <th>FB_ORIGIN</th>\n",
       "      <th>FB_COMPLEX_TYPE</th>\n",
       "      <th>FB_MULTIPLE_HAPS</th>\n",
       "      <th>FB_MULTIALLELIC</th>\n",
       "      <th>FB_REF_GL</th>\n",
       "      <th>FB_ALT_GL</th>\n",
       "      <th>predicted</th>\n",
       "      <th>pred_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>10518</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>other</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.883481</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>60.0</td>\n",
       "      <td>3.435650e-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-137.979</td>\n",
       "      <td>0.946788</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>40246</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>other</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.993000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1.514530e-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-181.604</td>\n",
       "      <td>0.974799</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag    POS REF ALT  GT REGION     FB_AF  COV_RATIO  \\\n",
       "0  sim1_h37rv_mutant_1_50_0.01  10518   T   A   0  other  0.043478   0.883481   \n",
       "1  sim1_h37rv_mutant_1_50_0.01  40246   C   T   0  other  0.033898   0.993000   \n",
       "\n",
       "   CLIPPED_BASES_RATIO  DISCORDANT_READS_RATIO  ...  FB_MQ       FB_QUAL  \\\n",
       "0                  0.0                     0.0  ...   60.0  3.435650e-15   \n",
       "1                  0.0                     0.0  ...   60.0  1.514530e-15   \n",
       "\n",
       "   FB_ORIGIN  FB_COMPLEX_TYPE  FB_MULTIPLE_HAPS  FB_MULTIALLELIC  FB_REF_GL  \\\n",
       "0        NaN                0                 0                0        0.0   \n",
       "1        NaN                0                 0                0        0.0   \n",
       "\n",
       "   FB_ALT_GL  predicted  pred_class  \n",
       "0   -137.979   0.946788           1  \n",
       "1   -181.604   0.974799           1  \n",
       "\n",
       "[2 rows x 26 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicted_MV_SNPs_file = f\"{parent_dir}/variant_SNP_predictions.csv\"\n",
    "predicted_MV_SNPs_df = pd.read_csv(predicted_MV_SNPs_file, low_memory=False)\n",
    "print(predicted_MV_SNPs_df.shape)\n",
    "predicted_MV_SNPs_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1cc39c92-4940-41ad-a587-fb6e73331aa7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "source_dir = f\"{parent_dir}/position_metrics\"\n",
    "output_dir = f\"{parent_dir}/SNPs\"\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    SNP_df = predicted_MV_SNPs_df[predicted_MV_SNPs_df.tag == tag]\n",
    "    SNP_df = SNP_df.drop(\"tag\", axis=1)\n",
    "\n",
    "    # label with exclusion filters\n",
    "    SNP_df[\"EXCL_DP\"] = (SNP_df.FB_DP < 5).astype(int)\n",
    "    SNP_df[\"EXCL_MQ\"] = (SNP_df.FB_MQ < 40).astype(int)\n",
    "    SNP_df[\"EXCL_SAO\"] = ((SNP_df.FB_AO_F < 2) | (SNP_df.FB_AO_R < 2)).astype(int)\n",
    "    \n",
    "    SNP_df.to_csv(f\"{output_dir}/{tag}.SNPs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a17836-f23b-4728-b72f-a68550debd6c",
   "metadata": {},
   "source": [
    "# BinoSNP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dbbbae86-e7e2-41d4-884d-c1c9150e9865",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BStags = pd.read_csv(\"./data/source/binosnp_data_tags.csv\", names=[\"tag\"]).tag.values\n",
    "len(BStags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "41abe65b-2773-4d2a-a0dd-0198844134c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = \"/n/scratch/users/s/sm624/FP_characteristics/BSdata\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edb399cb-84d0-4b87-9f8e-7fb6482eb678",
   "metadata": {},
   "source": [
    "## Filter variant summary file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b152cd0e-c018-4123-9dab-ea86b6d7a161",
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = f\"{parent_dir}/position_metrics\"\n",
    "output_dir = f\"{parent_dir}/SNPs\"\n",
    "\n",
    "count = 0\n",
    "for tag in BStags:\n",
    "\n",
    "    position_metric_df = pd.read_csv(f\"{source_dir}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "\n",
    "    # filter for only the variants found by FreeBayes\n",
    "    position_metric_df = position_metric_df[position_metric_df.FB_found == 1]\n",
    "\n",
    "    # extract necessary columns\n",
    "    _position_metric_df = position_metric_df[[\"POS\", \"REF\", \"ALT\", \"GT\", \"REGION\", \"FB_AF\", \"COV_RATIO\", \"CLIPPED_BASES_RATIO\", \"DISCORDANT_READS_RATIO\",\n",
    "                                              \"MEAN_ALT_BQ\", \"FB_STRAND_BIAS\", \"FB_AO\", \"FB_AO_F\", \"FB_AO_R\", \"FB_DP\", \"FB_MQ\", \"FB_QUAL\",\n",
    "                                              \"FB_ORIGIN\", \"FB_COMPLEX_TYPE\", \"FB_MULTIPLE_HAPS\", \"FB_MULTIALLELIC\", \"FB_REF_GL\", \"FB_ALT_GL\"]]\n",
    "\n",
    "    # filter SNPs\n",
    "    SNP_df = _position_metric_df[(_position_metric_df.REF.str.len() == 1) & (_position_metric_df.ALT.str.len() == 1)].copy(deep=True)\n",
    "\n",
    "    # rename columns for consistency with variables used by filtering model\n",
    "    SNP_df = SNP_df.rename({\"MEAN_ALT_BQ\":\"Mean_BQ_ALT_allele\",\n",
    "                            \"FB_STRAND_BIAS\": \"SAF_prop_deviation_from_half\"}, axis=1)\n",
    "\n",
    "    # already split MNPs\n",
    "\n",
    "    SNP_df.to_csv(f\"{output_dir}/{tag}.candidate_SNPs.csv\", index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43ff868f-46a5-4610-af64-b06b2e8b7ff2",
   "metadata": {},
   "source": [
    "## Make tag ID file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ba386665-91eb-4876-ad77-55abe3c165a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "tag_id_df = pd.DataFrame(columns=[\"tag\"])\n",
    "tag_id_df.loc[0, :] = 1\n",
    "tag_id_df.loc[0, \"tag\"] = \"dummy\"\n",
    "\n",
    "count = 0\n",
    "for tag in BStags:\n",
    "\n",
    "    lowAF_SNP_file = f\"{output_dir}/{tag}.candidate_SNPs.csv\"\n",
    "\n",
    "    lowAF_SNP_df = pd.read_csv(lowAF_SNP_file)\n",
    "\n",
    "    add_df = pd.DataFrame(columns=[\"tag\"])\n",
    "    add_df[\"tag\"] = [tag]*lowAF_SNP_df.shape[0]\n",
    "\n",
    "    tag_id_df = pd.concat([tag_id_df, add_df])\n",
    "    tag_id_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "tag_id_df.drop(0, inplace=True)\n",
    "tag_id_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7077d4b2-6758-4385-96ed-e4bbb62762d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15492, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mix-SR1a-rpoB531-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mix-SR1a-rpoB531-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tag\n",
       "0  Mix-SR1a-rpoB531-1\n",
       "1  Mix-SR1a-rpoB531-1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(tag_id_df.shape)\n",
    "tag_id_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "54c8ed61-af38-45c5-af6b-f7e6d44030e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "tag_id_df.to_csv(\"/n/scratch/users/s/sm624/BSdata_benchmarking_tag_ids.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11df5013-9319-4c8b-9616-17296a3483ef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
