{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "daee7ed8-ec38-4491-a443-27766e330000",
   "metadata": {},
   "source": [
    "# [SOURCE] FP CHARACTERISTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "567ff8af-5482-46a9-8351-dc93ae5cb120",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "from pandas.errors import EmptyDataError\n",
    "\n",
    "import sys\n",
    "sys.path.append(\"./scripts/modules\")\n",
    "\n",
    "from benchmarking_definitions import *\n",
    "from regions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4ca71620-2c65-47e1-9f2b-46e337645fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_lengths = {\"DR\": DR_region_length, \"HT\": HT_region_length, \"LM\": sim_LM_region_length}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ba9d3c4a-762a-4934-9826-5f3a77c1dc2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2500, 1000)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tagsRv = pd.read_csv(\"./data/source/FINAL_tag_list.csv\", names=[\"tag\"])[\"tag\"].values\n",
    "tagsL1234 = pd.read_csv(\"./data/source/FINAL_tag_list_L1234.csv\", names=[\"tag\"]).tag.values\n",
    "\n",
    "len(tagsRv), len(tagsL1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35f291a1-0d24-41cc-a507-ec00704a5290",
   "metadata": {},
   "outputs": [],
   "source": [
    "ISS_Rv_mutant_summary_source = \"/n/scratch/users/s/sm624/benchmarking/variant_summaries/ISS_H37Rv\"\n",
    "ISS_L1234_mutant_summary_source = \"/n/scratch/users/s/sm624/benchmarking/variant_summaries/ISS_L1234\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "878997f6-1a29-4db2-8756-f228ea169df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ISS_Rv_position_metrics_source = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_H37Rv/position_metrics\"\n",
    "ISS_L1234_position_metrics_source = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_L1234/position_metrics\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c71b19bd-f529-4e78-9ff8-9234c1dde724",
   "metadata": {},
   "source": [
    "# [done] L1234: FP at AF > 50%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "70ff478b-ecca-4051-8084-0b0e56c0840e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "num_FP_AFover50_df = pd.DataFrame(columns=[\"tag\", \"tool\", \"num_FP\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        FP_AFover50 = FP_df[FP_df[f\"{tool}_AF\"] > 0.50].shape[0]\n",
    "\n",
    "        num_FP_AFover50_df.loc[df_i] = [tag, tool, FP_AFover50]\n",
    "        df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4a5505c6-79e4-4291-a126-de37b467e114",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6000, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>FB</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>LF</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag tool  num_FP\n",
       "0  sim1_L1_mutant_50_0.01   FB      99\n",
       "1  sim1_L1_mutant_50_0.01   LF      81"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(num_FP_AFover50_df.shape)\n",
    "num_FP_AFover50_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "32b5266c-de30-4e04-90f6-49c8f8b3ff03",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_FP_AFover50_df.to_csv(\"./data/FP_characteristics/ISS_L1234_num_FP_AFover50.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cc8976d-f64c-479a-8e32-873210c11ec0",
   "metadata": {},
   "source": [
    "# [done] Percentage of FP calls that are SNPs out of all FPs (per tool)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b7427b0-f45a-4aa8-93be-af8347e68ebe",
   "metadata": {},
   "source": [
    "## H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7c35149c-0cef-49e0-bc1c-25a9b3e03df0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "Rv_FP_type_df = pd.DataFrame(columns=[\"tag\"] + [f\"{tool}_percent_SNP\" for tool in tools] + [f\"{tool}_total\" for tool in tools])\n",
    "Rv_FP_type_df[\"tag\"] = tagsRv\n",
    "Rv_FP_type_df = Rv_FP_type_df.set_index(\"tag\")\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "        tool_total = FP_df.shape[0]\n",
    "\n",
    "        SNP_FP_df = FP_df[(FP_df.REF.str.len() == 1) & (FP_df.ALT.str.len() == 1)]\n",
    "        tool_SNP_total = SNP_FP_df.shape[0]\n",
    "\n",
    "        if tool_total > 0:\n",
    "            percent_SNP = (tool_SNP_total/tool_total)*100\n",
    "        else:\n",
    "            percent_SNP = np.nan\n",
    "\n",
    "        Rv_FP_type_df.loc[tag, f\"{tool}_percent_SNP\"] = percent_SNP\n",
    "        Rv_FP_type_df.loc[tag, f\"{tool}_total\"] = tool_total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "52684404-9322-4b8b-a7d2-18d32b4ef3e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FB_percent_SNP</th>\n",
       "      <th>LF_percent_SNP</th>\n",
       "      <th>MT_percent_SNP</th>\n",
       "      <th>PL_percent_SNP</th>\n",
       "      <th>VD_percent_SNP</th>\n",
       "      <th>VS_percent_SNP</th>\n",
       "      <th>FB_total</th>\n",
       "      <th>LF_total</th>\n",
       "      <th>MT_total</th>\n",
       "      <th>PL_total</th>\n",
       "      <th>VD_total</th>\n",
       "      <th>VS_total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sim1_h37rv_mutant_1_50_0.01</th>\n",
       "      <td>96.551724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>95.483871</td>\n",
       "      <td>NaN</td>\n",
       "      <td>145</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>72</td>\n",
       "      <td>310</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim1_h37rv_mutant_2_50_0.01</th>\n",
       "      <td>94.354839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>95.289855</td>\n",
       "      <td>NaN</td>\n",
       "      <td>124</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>276</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim1_h37rv_mutant_3_50_0.01</th>\n",
       "      <td>96.240602</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>95.932203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>63</td>\n",
       "      <td>295</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim1_h37rv_mutant_4_50_0.01</th>\n",
       "      <td>96.875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>95.238095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>294</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim1_h37rv_mutant_5_50_0.01</th>\n",
       "      <td>97.014925</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>95.394737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>304</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            FB_percent_SNP LF_percent_SNP MT_percent_SNP  \\\n",
       "tag                                                                        \n",
       "sim1_h37rv_mutant_1_50_0.01      96.551724            NaN          100.0   \n",
       "sim1_h37rv_mutant_2_50_0.01      94.354839            NaN            NaN   \n",
       "sim1_h37rv_mutant_3_50_0.01      96.240602            NaN          100.0   \n",
       "sim1_h37rv_mutant_4_50_0.01         96.875            NaN          100.0   \n",
       "sim1_h37rv_mutant_5_50_0.01      97.014925            NaN          100.0   \n",
       "\n",
       "                            PL_percent_SNP VD_percent_SNP VS_percent_SNP  \\\n",
       "tag                                                                        \n",
       "sim1_h37rv_mutant_1_50_0.01          100.0      95.483871            NaN   \n",
       "sim1_h37rv_mutant_2_50_0.01          100.0      95.289855            NaN   \n",
       "sim1_h37rv_mutant_3_50_0.01          100.0      95.932203            NaN   \n",
       "sim1_h37rv_mutant_4_50_0.01          100.0      95.238095            NaN   \n",
       "sim1_h37rv_mutant_5_50_0.01          100.0      95.394737            NaN   \n",
       "\n",
       "                            FB_total LF_total MT_total PL_total VD_total  \\\n",
       "tag                                                                        \n",
       "sim1_h37rv_mutant_1_50_0.01      145        0        7       72      310   \n",
       "sim1_h37rv_mutant_2_50_0.01      124        0        0       72      276   \n",
       "sim1_h37rv_mutant_3_50_0.01      133        0       11       63      295   \n",
       "sim1_h37rv_mutant_4_50_0.01      128        0        1       68      294   \n",
       "sim1_h37rv_mutant_5_50_0.01      134        0        1       69      304   \n",
       "\n",
       "                            VS_total  \n",
       "tag                                   \n",
       "sim1_h37rv_mutant_1_50_0.01        0  \n",
       "sim1_h37rv_mutant_2_50_0.01        0  \n",
       "sim1_h37rv_mutant_3_50_0.01        0  \n",
       "sim1_h37rv_mutant_4_50_0.01        0  \n",
       "sim1_h37rv_mutant_5_50_0.01        0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Rv_FP_type_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e6533ae3-c08b-495c-a355-d0ca98ec99ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rv_FP_type_df.to_csv(\"./data/FP_characteristics/ISS_H37Rv_percent_SNP.csv\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a781d2fe-f57f-4919-94c2-4a102ee491e3",
   "metadata": {},
   "source": [
    "## L1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "69be8426-8a55-4486-88d9-98e5ccd9e325",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "L1234_FP_type_df = pd.DataFrame(columns=[\"tag\"] + [f\"{tool}_percent_SNP\" for tool in tools] + [f\"{tool}_total\" for tool in tools])\n",
    "L1234_FP_type_df[\"tag\"] = tagsL1234\n",
    "L1234_FP_type_df = L1234_FP_type_df.set_index(\"tag\")\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        tool_total = FP_df.shape[0]\n",
    "\n",
    "        SNP_FP_df = FP_df[(FP_df.REF.str.len() == 1) & (FP_df.ALT.str.len() == 1)]\n",
    "        tool_SNP_total = SNP_FP_df.shape[0]\n",
    "\n",
    "        percent_SNP = (tool_SNP_total/tool_total)*100\n",
    "\n",
    "        L1234_FP_type_df.loc[tag, f\"{tool}_percent_SNP\"] = percent_SNP\n",
    "        L1234_FP_type_df.loc[tag, f\"{tool}_total\"] = tool_total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "afed4195-2745-4845-a664-6fc3f879d080",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FB_percent_SNP</th>\n",
       "      <th>LF_percent_SNP</th>\n",
       "      <th>MT_percent_SNP</th>\n",
       "      <th>PL_percent_SNP</th>\n",
       "      <th>VD_percent_SNP</th>\n",
       "      <th>VS_percent_SNP</th>\n",
       "      <th>FB_total</th>\n",
       "      <th>LF_total</th>\n",
       "      <th>MT_total</th>\n",
       "      <th>PL_total</th>\n",
       "      <th>VD_total</th>\n",
       "      <th>VS_total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sim1_L1_mutant_50_0.01</th>\n",
       "      <td>96.412037</td>\n",
       "      <td>94.134078</td>\n",
       "      <td>78.441558</td>\n",
       "      <td>97.791165</td>\n",
       "      <td>85.988201</td>\n",
       "      <td>95.918367</td>\n",
       "      <td>864</td>\n",
       "      <td>358</td>\n",
       "      <td>385</td>\n",
       "      <td>498</td>\n",
       "      <td>678</td>\n",
       "      <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim1_L2_mutant_50_0.01</th>\n",
       "      <td>95.535714</td>\n",
       "      <td>91.156463</td>\n",
       "      <td>74.587459</td>\n",
       "      <td>96.983759</td>\n",
       "      <td>86.933798</td>\n",
       "      <td>91.836735</td>\n",
       "      <td>672</td>\n",
       "      <td>294</td>\n",
       "      <td>303</td>\n",
       "      <td>431</td>\n",
       "      <td>574</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim1_L3_mutant_50_0.01</th>\n",
       "      <td>96.446701</td>\n",
       "      <td>93.714286</td>\n",
       "      <td>70.0</td>\n",
       "      <td>97.555556</td>\n",
       "      <td>88.651316</td>\n",
       "      <td>94.42623</td>\n",
       "      <td>788</td>\n",
       "      <td>350</td>\n",
       "      <td>370</td>\n",
       "      <td>450</td>\n",
       "      <td>608</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim2_L1_mutant_50_0.01</th>\n",
       "      <td>96.40914</td>\n",
       "      <td>95.856354</td>\n",
       "      <td>78.39196</td>\n",
       "      <td>97.864078</td>\n",
       "      <td>86.248132</td>\n",
       "      <td>96.527778</td>\n",
       "      <td>919</td>\n",
       "      <td>362</td>\n",
       "      <td>398</td>\n",
       "      <td>515</td>\n",
       "      <td>669</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sim2_L2_mutant_50_0.01</th>\n",
       "      <td>95.135908</td>\n",
       "      <td>91.317365</td>\n",
       "      <td>70.819672</td>\n",
       "      <td>96.86747</td>\n",
       "      <td>87.438825</td>\n",
       "      <td>90.657439</td>\n",
       "      <td>699</td>\n",
       "      <td>334</td>\n",
       "      <td>305</td>\n",
       "      <td>415</td>\n",
       "      <td>613</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       FB_percent_SNP LF_percent_SNP MT_percent_SNP  \\\n",
       "tag                                                                   \n",
       "sim1_L1_mutant_50_0.01      96.412037      94.134078      78.441558   \n",
       "sim1_L2_mutant_50_0.01      95.535714      91.156463      74.587459   \n",
       "sim1_L3_mutant_50_0.01      96.446701      93.714286           70.0   \n",
       "sim2_L1_mutant_50_0.01       96.40914      95.856354       78.39196   \n",
       "sim2_L2_mutant_50_0.01      95.135908      91.317365      70.819672   \n",
       "\n",
       "                       PL_percent_SNP VD_percent_SNP VS_percent_SNP FB_total  \\\n",
       "tag                                                                            \n",
       "sim1_L1_mutant_50_0.01      97.791165      85.988201      95.918367      864   \n",
       "sim1_L2_mutant_50_0.01      96.983759      86.933798      91.836735      672   \n",
       "sim1_L3_mutant_50_0.01      97.555556      88.651316       94.42623      788   \n",
       "sim2_L1_mutant_50_0.01      97.864078      86.248132      96.527778      919   \n",
       "sim2_L2_mutant_50_0.01       96.86747      87.438825      90.657439      699   \n",
       "\n",
       "                       LF_total MT_total PL_total VD_total VS_total  \n",
       "tag                                                                  \n",
       "sim1_L1_mutant_50_0.01      358      385      498      678      294  \n",
       "sim1_L2_mutant_50_0.01      294      303      431      574      245  \n",
       "sim1_L3_mutant_50_0.01      350      370      450      608      305  \n",
       "sim2_L1_mutant_50_0.01      362      398      515      669      288  \n",
       "sim2_L2_mutant_50_0.01      334      305      415      613      289  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L1234_FP_type_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "797348d5-2be2-4d60-acf0-0ecd2f722140",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1234_FP_type_df.to_csv(\"./data/FP_characteristics/ISS_L1234_percent_SNP.csv\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a4e46fa-c482-4010-819a-6c16d3c66226",
   "metadata": {},
   "source": [
    "# [done] Genome-wide FPR in H37Rv vs L1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a2339462-7f8e-4ca3-a8ca-ae3924cd96e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900 3000 3100 3200 3300 3400 "
     ]
    }
   ],
   "source": [
    "genome_FPR_df = pd.DataFrame(columns=[\"tag\", \"genome\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "\n",
    "genome = \"H37Rv\"\n",
    "\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "        tool_FPR = tool_total_FP/genome_length\n",
    "\n",
    "        genome_FPR_df.loc[df_i] = [tag, genome, tool, tool_total_FP, tool_FPR]\n",
    "        df_i += 1\n",
    "\n",
    "genome = \"L1-4\"\n",
    "\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "        tool_FPR = tool_total_FP/genome_length\n",
    "\n",
    "        genome_FPR_df.loc[df_i] = [tag, genome, tool, tool_total_FP, tool_FPR]\n",
    "        df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1e7b3e8f-43c0-4937-87f9-c7450a456d01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21000, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>genome</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>FB</td>\n",
       "      <td>145</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>LF</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>MT</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>PL</td>\n",
       "      <td>72</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>VD</td>\n",
       "      <td>310</td>\n",
       "      <td>0.000070</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag genome tool  num_FP       FPR\n",
       "0  sim1_h37rv_mutant_1_50_0.01  H37Rv   FB     145  0.000033\n",
       "1  sim1_h37rv_mutant_1_50_0.01  H37Rv   LF       0  0.000000\n",
       "2  sim1_h37rv_mutant_1_50_0.01  H37Rv   MT       7  0.000002\n",
       "3  sim1_h37rv_mutant_1_50_0.01  H37Rv   PL      72  0.000016\n",
       "4  sim1_h37rv_mutant_1_50_0.01  H37Rv   VD     310  0.000070"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(genome_FPR_df.shape)\n",
    "genome_FPR_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e99f8cc-48ed-40f1-ab72-d1357d02cf79",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_FPR_df.to_csv(\"./data/FP_characteristics/ISS_genome_wide_FPR.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0aa59cd-6f95-40ef-a1dd-a0a2fd1b8b88",
   "metadata": {},
   "source": [
    "# [done] Percentage of total FP that are LM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "41b0409b-6d84-4514-a185-17cf3e6bc72b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900 3000 3100 3200 3300 3400 "
     ]
    }
   ],
   "source": [
    "percent_LM_FP_df = pd.DataFrame(columns=[\"tag\", \"genome\", \"tool\", \"num_FP\", \"num_LM_FP\", \"percent_LM_FP\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "\n",
    "genome = \"H37Rv\"\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "\n",
    "        LM_tool_total_FP = tool_FP_df[tool_FP_df.REGION == \"LM\"].shape[0]\n",
    "\n",
    "        if tool_total_FP > 0:\n",
    "            percent_LM_FP = (LM_tool_total_FP/tool_total_FP)*100            \n",
    "\n",
    "            percent_LM_FP_df.loc[df_i] = [tag, genome, tool, tool_total_FP, LM_tool_total_FP, percent_LM_FP]\n",
    "            df_i += 1\n",
    "\n",
    "genome = \"L1-4\"\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "\n",
    "        LM_tool_total_FP = tool_FP_df[tool_FP_df.REGION == \"LM\"].shape[0]\n",
    "\n",
    "        if tool_total_FP > 0:\n",
    "            percent_LM_FP = (LM_tool_total_FP/tool_total_FP)*100            \n",
    "\n",
    "            percent_LM_FP_df.loc[df_i] = [tag, genome, tool, tool_total_FP, LM_tool_total_FP, percent_LM_FP]\n",
    "            df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23ce127b-8c52-44a0-bd08-b5aca44ab6be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13921, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>genome</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>num_LM_FP</th>\n",
       "      <th>percent_LM_FP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>FB</td>\n",
       "      <td>145</td>\n",
       "      <td>12</td>\n",
       "      <td>8.275862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>MT</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>14.285714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag genome tool  num_FP  num_LM_FP  percent_LM_FP\n",
       "0  sim1_h37rv_mutant_1_50_0.01  H37Rv   FB     145         12       8.275862\n",
       "1  sim1_h37rv_mutant_1_50_0.01  H37Rv   MT       7          1      14.285714"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(percent_LM_FP_df.shape)\n",
    "percent_LM_FP_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b135dffa-3971-4928-90ba-0ab74d098982",
   "metadata": {},
   "outputs": [],
   "source": [
    "percent_LM_FP_df.to_csv(\"./data/FP_characteristics/ISS_percent_LM_FP.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17235667-fe25-4216-a5ba-46f7f04fde0c",
   "metadata": {},
   "source": [
    "# [done] Percentage of total FP at AF = 1%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fae0bdc5-672d-4159-8124-014840044bd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900 3000 3100 3200 3300 3400 "
     ]
    }
   ],
   "source": [
    "percent_AF1_FP_df = pd.DataFrame(columns=[\"tag\", \"genome\", \"tool\", \"num_FP\", \"num_AF1_FP\", \"percent_AF1_FP\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "\n",
    "genome = \"H37Rv\"\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "\n",
    "        AF1_tool_total_FP = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] == 0.01].shape[0]\n",
    "\n",
    "        if tool_total_FP > 0:\n",
    "            percent_AF1_FP = (AF1_tool_total_FP/tool_total_FP)*100            \n",
    "\n",
    "            percent_AF1_FP_df.loc[df_i] = [tag, genome, tool, tool_total_FP, AF1_tool_total_FP, percent_AF1_FP]\n",
    "            df_i += 1\n",
    "\n",
    "genome = \"L1-4\"\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "\n",
    "        AF1_tool_total_FP = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] == 0.01].shape[0]\n",
    "\n",
    "        if tool_total_FP > 0:\n",
    "            percent_AF1_FP = (AF1_tool_total_FP/tool_total_FP)*100            \n",
    "\n",
    "            percent_AF1_FP_df.loc[df_i] = [tag, genome, tool, tool_total_FP, AF1_tool_total_FP, percent_AF1_FP]\n",
    "            df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b20d7bcd-284a-4fe8-871b-c03c293ec615",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13921, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>genome</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>num_AF1_FP</th>\n",
       "      <th>percent_AF1_FP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>FB</td>\n",
       "      <td>145</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>MT</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag genome tool  num_FP  num_AF1_FP  percent_AF1_FP\n",
       "0  sim1_h37rv_mutant_1_50_0.01  H37Rv   FB     145           0             0.0\n",
       "1  sim1_h37rv_mutant_1_50_0.01  H37Rv   MT       7           0             0.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(percent_AF1_FP_df.shape)\n",
    "percent_AF1_FP_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "68b638cd-80f9-4f74-8ed0-f8de0d583ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "percent_AF1_FP_df.to_csv(\"./data/FP_characteristics/ISS_percent_AF1_FP.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b08b70-d0b2-4d26-82f7-088aff2c93e6",
   "metadata": {},
   "source": [
    "# [done] FP AF distribution per LM region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d6bf3a-5d4e-4de8-848d-e17100dcb319",
   "metadata": {},
   "source": [
    "## H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "49bb8830-2d71-419f-a82a-f14c839a0ce6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "region_AF_distr_dict = {t:{\"LM\":[], \"non-LM\":[]} for t in tools}\n",
    "\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "        LM_tool_FP_df = tool_FP_df[tool_FP_df.REGION == \"LM\"]\n",
    "        region_AF_distr_dict[tool][\"LM\"] += list(LM_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        nonLM_tool_FP_df = tool_FP_df[tool_FP_df.REGION != \"LM\"]\n",
    "        region_AF_distr_dict[tool][\"non-LM\"] += list(nonLM_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "665e2ae9-3185-4bfa-8c62-4575c7244faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_H37Rv_region_AF_distr.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(region_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46600ff1-42d3-4736-a9d2-604756e43e5b",
   "metadata": {},
   "source": [
    "## L1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "68de98b6-c905-4553-a7fa-d61bbc41bba1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "region_AF_distr_dict = {t:{\"LM\":[], \"non-LM\":[]} for t in tools}\n",
    "\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        LM_tool_FP_df = tool_FP_df[tool_FP_df.REGION == \"LM\"]\n",
    "        region_AF_distr_dict[tool][\"LM\"] += list(LM_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        nonLM_tool_FP_df = tool_FP_df[tool_FP_df.REGION != \"LM\"]\n",
    "        region_AF_distr_dict[tool][\"non-LM\"] += list(nonLM_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b4d35a06-010c-4c18-96e2-012c4d022f7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_L1234_region_AF_distr.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(region_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79ad4bed-8a0d-4850-8a15-aec305cb4bb3",
   "metadata": {},
   "source": [
    "# [done] FP AF distribution per DR region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80dfb76f-14c4-4747-b85c-2904d0fbe627",
   "metadata": {},
   "source": [
    "## H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a1fbe6f5-85db-41fe-92db-21e7acb31972",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "DR_region_AF_distr_dict = {t:{\"DR\":[], \"non-DR\":[]} for t in tools}\n",
    "\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "        DR_tool_FP_df = tool_FP_df[tool_FP_df.REGION == \"DR\"]\n",
    "        DR_region_AF_distr_dict[tool][\"DR\"] += list(DR_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        nonDR_tool_FP_df = tool_FP_df[tool_FP_df.REGION != \"DR\"]\n",
    "        DR_region_AF_distr_dict[tool][\"non-DR\"] += list(nonDR_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "37fe99a3-5837-4385-830e-9e104ca83394",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_H37Rv_DR_region_AF_distr.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(DR_region_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536fd81e-f7cb-4c49-a8cb-770650054503",
   "metadata": {},
   "source": [
    "## L1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "203004a4-e9da-4c38-a9bb-5c63723f8890",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "DR_region_AF_distr_dict = {t:{\"DR\":[], \"non-DR\":[]} for t in tools}\n",
    "\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        DR_tool_FP_df = tool_FP_df[tool_FP_df.REGION == \"DR\"]\n",
    "        DR_region_AF_distr_dict[tool][\"DR\"] += list(DR_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        nonDR_tool_FP_df = tool_FP_df[tool_FP_df.REGION != \"DR\"]\n",
    "        DR_region_AF_distr_dict[tool][\"non-DR\"] += list(nonDR_tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0f2e06ff-3e35-46c6-80e0-bfdfdc8c6a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_L1234_DR_region_AF_distr.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(DR_region_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e52c0a4-a53d-47ec-a644-19c9eb3b934f",
   "metadata": {},
   "source": [
    "# [done] FP AF distribution in DR vs LM vs OTHER regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5e4e52f8-e4f8-47e3-876b-e2824d2908a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS\n",
       "0    1\n",
       "1    3\n",
       "2    5\n",
       "3    7\n",
       "4   10\n",
       "5   12\n",
       "6   23"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_LM_pos = [1, 3, 5, 9, 11, 12]\n",
    "test_DR_pos = [1, 3, 7]\n",
    "test_df = pd.DataFrame()\n",
    "test_df[\"POS\"] = [1, 3, 5, 7, 10, 12, 23]\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3c314392-5e64-4edf-bf54-0aa5e2fe332d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS\n",
       "3    7"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DR_nonLM\n",
    "test_df[test_df.POS.isin(test_DR_pos) & ~test_df.POS.isin(test_LM_pos)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dcc486ad-ec02-40a5-adf9-f645f804a0c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS\n",
       "4   10\n",
       "6   23"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# other_nonLM\n",
    "test_df[~test_df.POS.isin(test_DR_pos) & ~test_df.POS.isin(test_LM_pos)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ca506f5f-836c-41ca-9ec7-ba2d85147882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS\n",
       "0    1\n",
       "1    3\n",
       "2    5\n",
       "5   12"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# LM\n",
    "test_df[test_df.POS.isin(test_LM_pos)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c1a6e42-9052-4467-9e59-ba460467f13a",
   "metadata": {},
   "source": [
    "## H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "19ec53b5-e281-4512-872a-224ba4030de9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "all_region_AF_distr_dict = {t:{\"DR_nonLM\":[], \"other_nonLM\":[], \"LM\": []} for t in tools}\n",
    "\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    FP_df = variant_summary_df[variant_summary_df.GT == 0]\n",
    "\n",
    "    DR_nonLM_FP_df = FP_df[FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    other_nonLM_FP_df = FP_df[~FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    LM_FP_df = FP_df[FP_df.POS.isin(LM_pos)]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_DR_nonLM_FP_df = DR_nonLM_FP_df[DR_nonLM_FP_df[f\"{tool}_found\"] == 1]\n",
    "        all_region_AF_distr_dict[tool][\"DR_nonLM\"] += list(tool_DR_nonLM_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        tool_other_nonLM_FP_df = other_nonLM_FP_df[other_nonLM_FP_df[f\"{tool}_found\"] == 1]\n",
    "        all_region_AF_distr_dict[tool][\"other_nonLM\"] += list(tool_other_nonLM_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        tool_LM_FP_df = LM_FP_df[LM_FP_df[f\"{tool}_found\"] == 1]\n",
    "        all_region_AF_distr_dict[tool][\"LM\"] += list(tool_LM_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b311c0b5-ccf2-4c24-ac21-c3101dbe27ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_H37Rv_FP_AF_distr_DR_LM_other.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(all_region_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77ffc993-23c7-4dff-9ec3-18c3079b8d16",
   "metadata": {},
   "source": [
    "## L1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6901e10d-3cc7-4f13-a4d3-d99edcfafe22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "all_region_AF_distr_dict = {t:{\"DR_nonLM\":[], \"other_nonLM\":[], \"LM\": []} for t in tools}\n",
    "\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    FP_df = variant_summary_df[(variant_summary_df.L_GT == 0) & (variant_summary_df.GT == 0)]\n",
    "\n",
    "    DR_nonLM_FP_df = FP_df[FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    other_nonLM_FP_df = FP_df[~FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    LM_FP_df = FP_df[FP_df.POS.isin(LM_pos)]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_DR_nonLM_FP_df = DR_nonLM_FP_df[(DR_nonLM_FP_df[f\"{tool}_found\"] == 1) & (DR_nonLM_FP_df[f\"{tool}_AF\"] <= 0.50)]\n",
    "        all_region_AF_distr_dict[tool][\"DR_nonLM\"] += list(tool_DR_nonLM_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        tool_other_nonLM_FP_df = other_nonLM_FP_df[(other_nonLM_FP_df[f\"{tool}_found\"] == 1) & (other_nonLM_FP_df[f\"{tool}_AF\"] <= 0.50)]\n",
    "        all_region_AF_distr_dict[tool][\"other_nonLM\"] += list(tool_other_nonLM_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n",
    "        tool_LM_FP_df = LM_FP_df[(LM_FP_df[f\"{tool}_found\"] == 1) & (LM_FP_df[f\"{tool}_AF\"] <= 0.50)]\n",
    "        all_region_AF_distr_dict[tool][\"LM\"] += list(tool_LM_FP_df[f\"{tool}_AF\"].values*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2b2229dd-7dab-43b7-8e08-2df495a67251",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_L1234_FP_AF_distr_DR_LM_other.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(all_region_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6c4a957-2aeb-48d1-bde9-18a105232339",
   "metadata": {},
   "source": [
    "# [done] FP AF distribution per depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8bfcfa4c-e049-4b58-a091-6a61a5422953",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900 3000 3100 3200 3300 3400 "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "depth_AF_distr_dict = {t:{d:[] for d in depths} for t in tools}\n",
    "\n",
    "for depth in depths:\n",
    "\n",
    "    for sim_i in range(1, 6):\n",
    "\n",
    "        for freq in freqs:\n",
    "\n",
    "            for mutant_num in range(1, 11):\n",
    "\n",
    "                if count % 100 == 0:\n",
    "                    print(count, end=\" \")\n",
    "                count += 1\n",
    "\n",
    "                tag = f\"sim{sim_i}_h37rv_mutant_{mutant_num}_{depth}_{freq}\"\n",
    "\n",
    "                variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "                for tool in tools:\n",
    "            \n",
    "                    tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    depth_AF_distr_dict[tool][depth] += list(tool_FP_df[f\"{tool}_AF\"].values*100)\n",
    "                    \n",
    "\n",
    "            for lineage in range(1, 5):\n",
    "\n",
    "                if count % 100 == 0:\n",
    "                    print(count, end=\" \")\n",
    "                count += 1\n",
    "\n",
    "                tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "\n",
    "                variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "                variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "            \n",
    "                for tool in tools:\n",
    "            \n",
    "                    tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "                    depth_AF_distr_dict[tool][depth] += list(tool_FP_df[f\"{tool}_AF\"].values*100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "141fc143-a704-4995-bbe3-da5f30ccd70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_depth_AF_distr_dict.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(depth_AF_distr_dict, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4faaf4e9-4ad9-484f-aef9-608b4fb7bd6f",
   "metadata": {},
   "source": [
    "# [done] FPR per depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8265ece-82d3-455f-8338-df16ad1b4900",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900 3000 3100 3200 3300 3400 "
     ]
    }
   ],
   "source": [
    "depth_FPR_df = pd.DataFrame(columns=[\"tag\", \"depth\", \"genome\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "\n",
    "genome = \"H37Rv\"\n",
    "\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    depth = int(tag.split(\"_\")[-2])\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "        tool_FPR = tool_total_FP/genome_length\n",
    "\n",
    "        depth_FPR_df.loc[df_i] = [tag, depth, genome, tool, tool_total_FP, tool_FPR]\n",
    "        df_i += 1\n",
    "\n",
    "genome = \"L1-4\"\n",
    "\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    depth = int(tag.split(\"_\")[-2])\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        tool_total_FP = tool_FP_df.shape[0]\n",
    "        tool_FPR = tool_total_FP/genome_length\n",
    "\n",
    "        depth_FPR_df.loc[df_i] = [tag, depth, genome, tool, tool_total_FP, tool_FPR]\n",
    "        df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9d188805-bbac-4af8-b890-04f8722edae2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21000, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>depth</th>\n",
       "      <th>genome</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>50</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>FB</td>\n",
       "      <td>145</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>50</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>LF</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag  depth genome tool  num_FP       FPR\n",
       "0  sim1_h37rv_mutant_1_50_0.01     50  H37Rv   FB     145  0.000033\n",
       "1  sim1_h37rv_mutant_1_50_0.01     50  H37Rv   LF       0  0.000000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(depth_FPR_df.shape)\n",
    "depth_FPR_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "65c37a1d-a9fd-4903-af80-4af4243e438e",
   "metadata": {},
   "outputs": [],
   "source": [
    "depth_FPR_df.to_csv(\"./data/FP_characteristics/ISS_FPR_by_depth.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2142c2d-0c82-45fe-89ec-04add7847aca",
   "metadata": {},
   "source": [
    "# [done] L1234: FPR per region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0a217c29-0283-4ed0-aef0-d3f33b4dc323",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_region_lengths = {\"DR\": DR_region_length,\n",
    "                      \"HT\": HT_region_length,\n",
    "                      \"LM\": LM_region_length,\n",
    "                      \"other\": genome_length-DR_region_length-HT_region_length-LM_region_length}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "76c51258-c738-4c34-9468-1b8fc52b05a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "L1234_region_FPR_df = pd.DataFrame(columns=[\"tag\", \"region\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        for region in regions_other:\n",
    "\n",
    "            region_tool_num_FP = tool_FP_df[tool_FP_df.REGION == region].shape[0]\n",
    "\n",
    "            region_length = all_region_lengths[region]\n",
    "\n",
    "            region_tool_FPR = region_tool_num_FP/region_length\n",
    "\n",
    "            L1234_region_FPR_df.loc[df_i] = [tag, region, tool, region_tool_num_FP, region_tool_FPR]\n",
    "            df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2f34d200-3be8-460b-b474-ee844620f5a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24000, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>region</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>DR</td>\n",
       "      <td>FB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>HT</td>\n",
       "      <td>FB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag region tool  num_FP  FPR\n",
       "0  sim1_L1_mutant_50_0.01     DR   FB       0  0.0\n",
       "1  sim1_L1_mutant_50_0.01     HT   FB       0  0.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(L1234_region_FPR_df.shape)\n",
    "L1234_region_FPR_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "41f80788-1d29-4122-8340-a3f929d1c857",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1234_region_FPR_df.to_csv(\"./data/FP_characteristics/ISS_L1234_region_FPR.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e12be1d-1178-4bfe-87e4-20a6673e396e",
   "metadata": {},
   "source": [
    "# [done] L1234: FPR in DR vs LM vs OTHER regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c34cbd58-e41e-4655-98e0-089b5cd3b2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "DR_pos_arr = np.array(list(DR_pos.keys()))\n",
    "LM_pos_arr = np.array(LM_pos)\n",
    "\n",
    "DR_or_LM_pos_arr = np.union1d(DR_pos_arr, LM_pos_arr)\n",
    "genome_pos_arr = np.arange(1, genome_length+1)\n",
    "\n",
    "DR_nonLM_pos = DR_pos_arr[~np.isin(DR_pos_arr, LM_pos_arr)]\n",
    "other_nonLM_pos = genome_pos_arr[~np.isin(genome_pos_arr, DR_or_LM_pos_arr)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "eae01801-34dc-4332-bed0-95d9647f9c87",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_region_lengths = {\"DR_nonLM\": len(DR_nonLM_pos),\n",
    "                      \"other_nonLM\": len(other_nonLM_pos),\n",
    "                      \"LM\": len(LM_pos)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b948ab1f-cd46-40ff-a0ad-2f97811b4cdb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "L1234_region_FPR_df = pd.DataFrame(columns=[\"tag\", \"region\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    FP_df = variant_summary_df[(variant_summary_df.L_GT == 0) & (variant_summary_df.GT == 0)]\n",
    "\n",
    "    DR_nonLM_FP_df = FP_df[FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    other_nonLM_FP_df = FP_df[~FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    LM_FP_df = FP_df[FP_df.POS.isin(LM_pos)]\n",
    "\n",
    "    region_FP_dfs = {\"DR_nonLM\": DR_nonLM_FP_df,\n",
    "                  \"other_nonLM\": other_nonLM_FP_df,\n",
    "                  \"LM\": LM_FP_df}\n",
    "\n",
    "    for region in region_FP_dfs:\n",
    "\n",
    "        region_FP_df = region_FP_dfs[region]\n",
    "\n",
    "        region_length = all_region_lengths[region]\n",
    "\n",
    "        for tool in tools:\n",
    "\n",
    "            tool_region_FP_df = region_FP_df[(region_FP_df[f\"{tool}_found\"] == 1) & (region_FP_df[f\"{tool}_AF\"] <= 0.50)]\n",
    "\n",
    "            tool_region_num_FP = tool_region_FP_df.shape[0]\n",
    "\n",
    "            tool_region_FPR = tool_region_num_FP/region_length\n",
    "\n",
    "            L1234_region_FPR_df.loc[df_i] = [tag, region, tool, tool_region_num_FP, tool_region_FPR]\n",
    "            df_i += 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6578358c-8f45-463e-a6e3-43a7da675596",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18000, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>region</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>DR_nonLM</td>\n",
       "      <td>FB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>DR_nonLM</td>\n",
       "      <td>LF</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag    region tool  num_FP  FPR\n",
       "0  sim1_L1_mutant_50_0.01  DR_nonLM   FB       0  0.0\n",
       "1  sim1_L1_mutant_50_0.01  DR_nonLM   LF       0  0.0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(L1234_region_FPR_df.shape)\n",
    "L1234_region_FPR_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "49cbeaa2-7f80-4372-998e-1cc351f56ab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1234_region_FPR_df.to_csv(\"./data/FP_characteristics/ISS_L1234_FPR_DR_LM_other.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b79fa674-839d-4505-af80-af867ed9e499",
   "metadata": {},
   "source": [
    "# [done] H37Rv: FPR in DR vs LM vs OTHER regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cfc4527b-4743-4c36-97a0-543e1c79bfb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "DR_pos_arr = np.array(list(DR_pos.keys()))\n",
    "LM_pos_arr = np.array(LM_pos)\n",
    "\n",
    "DR_or_LM_pos_arr = np.union1d(DR_pos_arr, LM_pos_arr)\n",
    "genome_pos_arr = np.arange(1, genome_length+1)\n",
    "\n",
    "DR_nonLM_pos = DR_pos_arr[~np.isin(DR_pos_arr, LM_pos_arr)]\n",
    "other_nonLM_pos = genome_pos_arr[~np.isin(genome_pos_arr, DR_or_LM_pos_arr)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4abda7c9-772a-43c4-bd4a-6bd68c1f4d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_region_lengths = {\"DR_nonLM\": len(DR_nonLM_pos),\n",
    "                      \"other_nonLM\": len(other_nonLM_pos),\n",
    "                      \"LM\": len(LM_pos)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8ce77045-4057-4b5d-b07f-c78d478ba729",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 "
     ]
    }
   ],
   "source": [
    "Rv_region_FPR_df = pd.DataFrame(columns=[\"tag\", \"region\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    FP_df = variant_summary_df[variant_summary_df.GT == 0]\n",
    "\n",
    "    DR_nonLM_FP_df = FP_df[FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    other_nonLM_FP_df = FP_df[~FP_df.POS.isin(DR_pos) & ~FP_df.POS.isin(LM_pos)]\n",
    "    LM_FP_df = FP_df[FP_df.POS.isin(LM_pos)]\n",
    "\n",
    "    region_FP_dfs = {\"DR_nonLM\": DR_nonLM_FP_df,\n",
    "                  \"other_nonLM\": other_nonLM_FP_df,\n",
    "                  \"LM\": LM_FP_df}\n",
    "\n",
    "    for region in region_FP_dfs:\n",
    "\n",
    "        region_FP_df = region_FP_dfs[region]\n",
    "\n",
    "        region_length = all_region_lengths[region]\n",
    "\n",
    "        for tool in tools:\n",
    "\n",
    "            tool_region_FP_df = region_FP_df[region_FP_df[f\"{tool}_found\"] == 1]\n",
    "\n",
    "            tool_region_num_FP = tool_region_FP_df.shape[0]\n",
    "\n",
    "            tool_region_FPR = tool_region_num_FP/region_length\n",
    "\n",
    "            Rv_region_FPR_df.loc[df_i] = [tag, region, tool, tool_region_num_FP, tool_region_FPR]\n",
    "            df_i += 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "016da65c-d279-464e-8195-0ca364137547",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(45000, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>region</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>DR_nonLM</td>\n",
       "      <td>FB</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>DR_nonLM</td>\n",
       "      <td>LF</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag    region tool  num_FP       FPR\n",
       "0  sim1_h37rv_mutant_1_50_0.01  DR_nonLM   FB       1  0.000027\n",
       "1  sim1_h37rv_mutant_1_50_0.01  DR_nonLM   LF       0  0.000000"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(Rv_region_FPR_df.shape)\n",
    "Rv_region_FPR_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6d148bb4-3d9e-4df2-a57f-376510541356",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rv_region_FPR_df.to_csv(\"./data/FP_characteristics/ISS_H37Rv_FPR_DR_LM_other.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "604f61cd-ad9d-4aa4-998f-523117aec906",
   "metadata": {},
   "source": [
    "# [done] L1234: FPR per region at AF  5%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c02dfe93-74d9-46d5-ab36-e508d1f6256f",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_region_lengths = {\"DR\": DR_region_length,\n",
    "                      \"HT\": HT_region_length,\n",
    "                      \"LM\": LM_region_length,\n",
    "                      \"other\": genome_length-DR_region_length-HT_region_length-LM_region_length}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "104d84f8-750d-43aa-9780-dbb674293c2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "L1234_region_FPR_AF5andOver_df = pd.DataFrame(columns=[\"tag\", \"region\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[(tool_FP_df[f\"{tool}_AF\"] >= 0.05) & (tool_FP_df[f\"{tool}_AF\"] <= 0.50)]\n",
    "\n",
    "        for region in regions_other:\n",
    "\n",
    "            region_tool_num_FP = tool_FP_df[tool_FP_df.REGION == region].shape[0]\n",
    "\n",
    "            region_length = all_region_lengths[region]\n",
    "\n",
    "            region_tool_FPR = region_tool_num_FP/region_length\n",
    "\n",
    "            L1234_region_FPR_AF5andOver_df.loc[df_i] = [tag, region, tool, region_tool_num_FP, region_tool_FPR]\n",
    "            df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "daccda63-ad74-4d27-af5a-4785773c128f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24000, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>region</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>DR</td>\n",
       "      <td>FB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>HT</td>\n",
       "      <td>FB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag region tool  num_FP  FPR\n",
       "0  sim1_L1_mutant_50_0.01     DR   FB       0  0.0\n",
       "1  sim1_L1_mutant_50_0.01     HT   FB       0  0.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(L1234_region_FPR_AF5andOver_df.shape)\n",
    "L1234_region_FPR_AF5andOver_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7eac2ff0-2805-40d9-8649-85fc209a0435",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1234_region_FPR_AF5andOver_df.to_csv(\"./data/FP_characteristics/ISS_L1234_region_AF5andOver_FPR.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9d3347-94d0-4ad3-8896-95dde37ba5b5",
   "metadata": {},
   "source": [
    "# [done] Percentage of FP with FP < 5% and FP  5% per LM region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7673684a-2c96-4b2f-9d3b-8ff9a3c02d38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 1000 1100 1200 1300 1400 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900 3000 3100 3200 3300 3400 "
     ]
    }
   ],
   "source": [
    "AF5_tool_FP_stats_df = pd.DataFrame(columns=[\"tag\", \"genome\", \"tool\", \"LM_tool_FP\", \"LM_tool_FP_underAF5\", \"LM_tool_FP_AF5andOver\",\n",
    "                                             \"nonLM_tool_FP\", \"nonLM_tool_FP_underAF5\", \"nonLM_tool_FP_AF5andOver\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "\n",
    "genome = \"H37Rv\"\n",
    "\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    depth = int(tag.split(\"_\")[-2])\n",
    "\n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_Rv_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "        LM_tool_FP_df = tool_FP_df[tool_FP_df.REGION == \"LM\"]\n",
    "        LM_tool_FP = LM_tool_FP_df.shape[0]\n",
    "        LM_tool_FP_underAF5 = LM_tool_FP_df[LM_tool_FP_df[f\"{tool}_AF\"] < 0.05].shape[0]\n",
    "        LM_tool_FP_AF5andOver = LM_tool_FP_df[LM_tool_FP_df[f\"{tool}_AF\"] >= 0.05].shape[0]\n",
    "\n",
    "        nonLM_tool_FP_df = tool_FP_df[tool_FP_df.REGION != \"LM\"]\n",
    "        nonLM_tool_FP = nonLM_tool_FP_df.shape[0]\n",
    "        nonLM_tool_FP_underAF5 = nonLM_tool_FP_df[nonLM_tool_FP_df[f\"{tool}_AF\"] < 0.05].shape[0]\n",
    "        nonLM_tool_FP_AF5andOver = nonLM_tool_FP_df[nonLM_tool_FP_df[f\"{tool}_AF\"] >= 0.05].shape[0]\n",
    "        \n",
    "        AF5_tool_FP_stats_df.loc[df_i] = [tag, genome, tool, LM_tool_FP, LM_tool_FP_underAF5, LM_tool_FP_AF5andOver,\n",
    "                                          nonLM_tool_FP, nonLM_tool_FP_underAF5, nonLM_tool_FP_AF5andOver]\n",
    "        df_i += 1\n",
    "\n",
    "genome = \"L1-4\"\n",
    "\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    depth = int(tag.split(\"_\")[-2])\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "        \n",
    "        LM_tool_FP_df = tool_FP_df[tool_FP_df.REGION == \"LM\"]\n",
    "        LM_tool_FP = LM_tool_FP_df.shape[0]\n",
    "        LM_tool_FP_underAF5 = LM_tool_FP_df[LM_tool_FP_df[f\"{tool}_AF\"] < 0.05].shape[0]\n",
    "        LM_tool_FP_AF5andOver = LM_tool_FP_df[LM_tool_FP_df[f\"{tool}_AF\"] >= 0.05].shape[0]\n",
    "\n",
    "        nonLM_tool_FP_df = tool_FP_df[tool_FP_df.REGION != \"LM\"]\n",
    "        nonLM_tool_FP = nonLM_tool_FP_df.shape[0]\n",
    "        nonLM_tool_FP_underAF5 = nonLM_tool_FP_df[nonLM_tool_FP_df[f\"{tool}_AF\"] < 0.05].shape[0]\n",
    "        nonLM_tool_FP_AF5andOver = nonLM_tool_FP_df[nonLM_tool_FP_df[f\"{tool}_AF\"] >= 0.05].shape[0]\n",
    "        \n",
    "        AF5_tool_FP_stats_df.loc[df_i] = [tag, genome, tool, LM_tool_FP, LM_tool_FP_underAF5, LM_tool_FP_AF5andOver,\n",
    "                                          nonLM_tool_FP, nonLM_tool_FP_underAF5, nonLM_tool_FP_AF5andOver]\n",
    "        df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ec73edc9-55ef-49ae-ad71-e77b7f996a13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21000, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>genome</th>\n",
       "      <th>tool</th>\n",
       "      <th>LM_tool_FP</th>\n",
       "      <th>LM_tool_FP_underAF5</th>\n",
       "      <th>LM_tool_FP_AF5andOver</th>\n",
       "      <th>nonLM_tool_FP</th>\n",
       "      <th>nonLM_tool_FP_underAF5</th>\n",
       "      <th>nonLM_tool_FP_AF5andOver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>FB</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>132</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>LF</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>PL</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>49</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sim1_h37rv_mutant_1_50_0.01</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>VD</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>288</td>\n",
       "      <td>287</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag genome tool  LM_tool_FP  LM_tool_FP_underAF5  \\\n",
       "0  sim1_h37rv_mutant_1_50_0.01  H37Rv   FB          12                   12   \n",
       "1  sim1_h37rv_mutant_1_50_0.01  H37Rv   LF           0                    0   \n",
       "2  sim1_h37rv_mutant_1_50_0.01  H37Rv   MT           1                    0   \n",
       "3  sim1_h37rv_mutant_1_50_0.01  H37Rv   PL           6                    5   \n",
       "4  sim1_h37rv_mutant_1_50_0.01  H37Rv   VD          22                   22   \n",
       "\n",
       "   LM_tool_FP_AF5andOver  nonLM_tool_FP  nonLM_tool_FP_underAF5  \\\n",
       "0                      0            133                     132   \n",
       "1                      0              0                       0   \n",
       "2                      1              6                       1   \n",
       "3                      1             66                      49   \n",
       "4                      0            288                     287   \n",
       "\n",
       "   nonLM_tool_FP_AF5andOver  \n",
       "0                         1  \n",
       "1                         0  \n",
       "2                         5  \n",
       "3                        17  \n",
       "4                         1  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(AF5_tool_FP_stats_df.shape)\n",
    "AF5_tool_FP_stats_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7497e2ae-18bc-4e65-84b6-8cebaef91101",
   "metadata": {},
   "outputs": [],
   "source": [
    "AF5_tool_FP_stats_df.to_csv(\"./data/FP_characteristics/ISS_AF5_tool_FP_stats.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "990807c5-42d8-4b1d-92a8-b4d8f5cdc895",
   "metadata": {},
   "source": [
    "# [done] FPR by lineage background L1-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c9e4e782-7581-4a26-a561-6c032a359129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100 200 300 400 500 600 700 800 900 "
     ]
    }
   ],
   "source": [
    "L1234_lineage_FPR_df = pd.DataFrame(columns=[\"tag\", \"lineage\", \"tool\", \"num_FP\", \"FPR\"])\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsL1234:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    lineage = \"Lineage \" + tag.split(\"_\")[1].strip(\"L\")\n",
    "    \n",
    "    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "\n",
    "    for tool in tools:\n",
    "\n",
    "        tool_FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "        tool_FP_df = tool_FP_df[tool_FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "\n",
    "        tool_num_FP = tool_FP_df.shape[0]\n",
    "\n",
    "        tool_FPR = tool_num_FP/genome_length\n",
    "\n",
    "        L1234_lineage_FPR_df.loc[df_i] = [tag, lineage, tool, tool_num_FP, tool_FPR]\n",
    "        df_i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ea3a4314-a040-4dc6-b1a5-d13ef51e8a27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6000, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>lineage</th>\n",
       "      <th>tool</th>\n",
       "      <th>num_FP</th>\n",
       "      <th>FPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>Lineage 1</td>\n",
       "      <td>FB</td>\n",
       "      <td>864</td>\n",
       "      <td>0.000196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim1_L1_mutant_50_0.01</td>\n",
       "      <td>Lineage 1</td>\n",
       "      <td>LF</td>\n",
       "      <td>358</td>\n",
       "      <td>0.000081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tag    lineage tool  num_FP       FPR\n",
       "0  sim1_L1_mutant_50_0.01  Lineage 1   FB     864  0.000196\n",
       "1  sim1_L1_mutant_50_0.01  Lineage 1   LF     358  0.000081"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(L1234_lineage_FPR_df.shape)\n",
    "L1234_lineage_FPR_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0a25e349-bf2d-49de-af1b-6137d7222be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "L1234_lineage_FPR_df.to_csv(\"./data/FP_characteristics/ISS_L1234_lineage_FPR.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92175be9-edd8-470a-9cc8-dcc56be885ae",
   "metadata": {},
   "source": [
    "# [done] L1234: SNV error model characteristics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "521cd2fd-e741-44b8-84b3-c49b9664d63c",
   "metadata": {},
   "source": [
    "## Coverage distribution (by simulated coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2ef1f57c-a098-481b-81e7-b042a1859b5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeBayes\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "LoFreq\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Mutect2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Pilon\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarDict\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarScan2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "COV_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "    \n",
    "for tool in tools:\n",
    "\n",
    "    print(tool_mapping[tool])\n",
    "\n",
    "    count = 0\n",
    "    for sim_i in range(1, 6):\n",
    "        for lineage in range(1, 5):\n",
    "            for depth in depths:\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_position_metrics_source}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "            \n",
    "                    COV_distr[tool][\"TP\"] += list(TP_df[\"COV_RATIO\"].values)\n",
    "                    COV_distr[tool][\"FP\"] += list(FP_df[\"COV_RATIO\"].values)\n",
    "\n",
    "    print()\n",
    "\n",
    "print()\n",
    "\n",
    "with open(f\"./data/FP_characteristics/ISS_L1234_COV_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(COV_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f028ef21-a4e4-4352-88fa-f820cebf29d8",
   "metadata": {},
   "source": [
    "## Strand bias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a91c5010-337f-464a-a5cb-8cd7783678fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "SB_tools = [\"FB\", \"MT\", \"VD\", \"VS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7a8d7b8f-916d-4d09-9279-645ee5f91bc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeBayes\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Mutect2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarDict\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarScan2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "SB_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "\n",
    "for tool in SB_tools:\n",
    "\n",
    "    print(tool_mapping[tool])\n",
    "\n",
    "    count = 0\n",
    "    for sim_i in range(1, 6):\n",
    "        for lineage in range(1, 5):\n",
    "            for depth in depths:\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_position_metrics_source}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "            \n",
    "                    SB_distr[tool][\"TP\"] += list(TP_df[f\"{tool}_STRAND_BIAS\"].values)\n",
    "                    SB_distr[tool][\"FP\"] += list(FP_df[f\"{tool}_STRAND_BIAS\"].values)\n",
    "\n",
    "    print()\n",
    "\n",
    "print()\n",
    "\n",
    "with open(f\"./data/FP_characteristics/ISS_L1234_SB_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(SB_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6189c6de-d061-47d3-9057-cb1a16863c3f",
   "metadata": {},
   "source": [
    "## Variant base quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b414e3df-53b8-405c-aa34-83ad2dfd1e61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeBayes\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "LoFreq\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Mutect2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Pilon\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarDict\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarScan2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "vBQ_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "\n",
    "for tool in tools:\n",
    "\n",
    "    print(tool_mapping[tool])\n",
    "\n",
    "    count = 0\n",
    "    for sim_i in range(1, 6):\n",
    "        for lineage in range(1, 5):\n",
    "            for depth in depths:\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_position_metrics_source}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "            \n",
    "                    vBQ_distr[tool][\"TP\"] += list(TP_df[\"MEAN_ALT_BQ\"].values)\n",
    "                    vBQ_distr[tool][\"FP\"] += list(FP_df[\"MEAN_ALT_BQ\"].values)\n",
    "\n",
    "    print()\n",
    "\n",
    "print()\n",
    "\n",
    "with open(f\"./data/FP_characteristics/ISS_L1234_vBQ_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(vBQ_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52b37783-013f-471c-849d-c26c64932df6",
   "metadata": {},
   "source": [
    "## Soft-clipped bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9adaf2d8-221d-4f24-9a2d-568c726b979b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeBayes\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "LoFreq\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Mutect2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Pilon\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarDict\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarScan2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "SC_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "\n",
    "for tool in tools:\n",
    "\n",
    "    print(tool_mapping[tool])\n",
    "\n",
    "    count = 0\n",
    "    for sim_i in range(1, 6):\n",
    "        for lineage in range(1, 5):\n",
    "            for depth in depths:\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_position_metrics_source}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "            \n",
    "                    SC_distr[tool][\"TP\"] += list(TP_df[\"CLIPPED_BASES_RATIO\"].values)\n",
    "                    SC_distr[tool][\"FP\"] += list(FP_df[\"CLIPPED_BASES_RATIO\"].values)\n",
    "\n",
    "    print()\n",
    "\n",
    "print()\n",
    "\n",
    "with open(f\"./data/FP_characteristics/ISS_L1234_SC_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(SC_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d17df650-e7b2-47b1-a097-f002976f8af6",
   "metadata": {},
   "source": [
    "## Discordant reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b96085f1-1215-4b49-8f1b-f55edec647e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeBayes\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "LoFreq\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Mutect2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Pilon\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarDict\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarScan2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "DR_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "\n",
    "for tool in tools:\n",
    "\n",
    "    print(tool_mapping[tool])\n",
    "\n",
    "    count = 0\n",
    "    for sim_i in range(1, 6):\n",
    "        for lineage in range(1, 5):\n",
    "            for depth in depths:\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_position_metrics_source}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "            \n",
    "                    DR_distr[tool][\"TP\"] += list(TP_df[\"DISCORDANT_READS_RATIO\"].values)\n",
    "                    DR_distr[tool][\"FP\"] += list(FP_df[\"DISCORDANT_READS_RATIO\"].values)\n",
    "\n",
    "    print()\n",
    "\n",
    "print()\n",
    "\n",
    "with open(f\"./data/FP_characteristics/ISS_L1234_DR_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(DR_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa946d63-d769-4b76-9253-f371a74cb1ad",
   "metadata": {},
   "source": [
    "## Mapping quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "76e36ba1-5ff0-411d-af83-438f97c3c8a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "MQ_tools = [\"FB\", \"MT\", \"PL\", \"VD\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b2027ec4-af03-411b-ae9f-1047137f69ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeBayes\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Mutect2\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "Pilon\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "VarDict\n",
      "0 100 200 300 400 500 600 700 800 900 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "MQ_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "\n",
    "for tool in MQ_tools:\n",
    "\n",
    "    print(tool_mapping[tool])\n",
    "\n",
    "    count = 0\n",
    "    for sim_i in range(1, 6):\n",
    "        for lineage in range(1, 5):\n",
    "            for depth in depths:\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_position_metrics_source}/{tag}.position_metrics.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = FP_df[FP_df[f\"{tool}_AF\"] <= 0.50]\n",
    "            \n",
    "                    MQ_distr[tool][\"TP\"] += list(TP_df[f\"{tool}_MQ\"].values)\n",
    "                    MQ_distr[tool][\"FP\"] += list(FP_df[f\"{tool}_MQ\"].values)\n",
    "\n",
    "    print()\n",
    "\n",
    "print()\n",
    "\n",
    "with open(f\"./data/FP_characteristics/ISS_L1234_MQ_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(MQ_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f01f02b-69ea-4b68-b53c-7738ff5a852f",
   "metadata": {},
   "source": [
    "## (dropped) AO distribution (by simulated coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a52576-767f-4c06-96c5-6817da9b9e3c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for depth in depths:\n",
    "\n",
    "    print(f\"{depth}x\")\n",
    "\n",
    "    depth_AO_distr = {t:{\"TP\":[], \"FP\":[]} for t in tools}\n",
    "    \n",
    "    for tool in tools:\n",
    "    \n",
    "        print(tool_mapping[tool])\n",
    "    \n",
    "        count = 0\n",
    "        for sim_i in range(1, 6):\n",
    "            for lineage in range(1, 5):\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    TP_df = variant_summary_df[(variant_summary_df.GT == 1) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "            \n",
    "                    depth_AO_distr[tool][\"TP\"] += list(TP_df[f\"{tool}_AO\"].values)\n",
    "                    depth_AO_distr[tool][\"FP\"] += list(FP_df[f\"{tool}_AO\"].values)\n",
    "    \n",
    "        print()\n",
    "\n",
    "    print()\n",
    "\n",
    "    with open(f\"./data/FP_characteristics/ISS_L1234_AO_distribution_{depth}x.pkl\", \"wb\") as out_f:\n",
    "        pickle.dump(depth_AO_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "436ea5ac-3280-4738-af53-6c17110d4c8d",
   "metadata": {},
   "source": [
    "## (dropped) FP AF distribution (by simulated coverage and region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70da48e0-1e93-4fb6-a4b1-c96860194af3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for depth in depths:\n",
    "\n",
    "    print(f\"{depth}x\")\n",
    "\n",
    "    depth_AF_distr = {t:{r:[] for r in regions_other} for t in tools}\n",
    "    \n",
    "    for tool in tools:\n",
    "    \n",
    "        print(tool_mapping[tool])\n",
    "    \n",
    "        count = 0\n",
    "        for sim_i in range(1, 6):\n",
    "            for lineage in range(1, 5):\n",
    "                for freq in freqs:\n",
    "    \n",
    "                    tag = f\"sim{sim_i}_L{lineage}_mutant_{depth}_{freq}\"\n",
    "    \n",
    "                    if count % 100 == 0:\n",
    "                        print(count, end=\" \")\n",
    "                    count += 1\n",
    "            \n",
    "                    variant_summary_df = pd.read_csv(f\"{ISS_L1234_mutant_summary_source}/{tag}.variant_summary.csv\", low_memory=False)\n",
    "                    variant_summary_df = variant_summary_df[variant_summary_df.L_GT == 0]\n",
    "                    \n",
    "                    FP_df = variant_summary_df[(variant_summary_df.GT == 0) & (variant_summary_df[f\"{tool}_found\"] == 1)]\n",
    "\n",
    "                    for region in regions_other:\n",
    "                        region_df = FP_df[FP_df.REGION == region]\n",
    "                        depth_AF_distr[tool][region] += list(region_df[f\"{tool}_AF\"].values)\n",
    "    \n",
    "        print()\n",
    "\n",
    "    print()\n",
    "\n",
    "    with open(f\"./data/FP_characteristics/ISS_L1234_FP_AF_distribution_by_region_{depth}x.pkl\", \"wb\") as out_f:\n",
    "        pickle.dump(depth_AF_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68d986b9-11bd-4e3b-913c-417584d0f3d5",
   "metadata": {},
   "source": [
    "# INDELS (FB only)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f70d90-9dc8-4d54-970e-4a3679524e96",
   "metadata": {},
   "source": [
    "## H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c789ceb4-35c1-423a-b9c1-08069ffee289",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rv_parent_dir = \"/n/scratch/users/s/sm624/FP_characteristics/ISS_H37Rv\"\n",
    "Rv_INDEL_source_dir = f\"{Rv_parent_dir}/INDELs\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "576b2c85-528f-4eab-b7db-6a5aeec28329",
   "metadata": {},
   "source": [
    "### Depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b320d5bd-18e1-4fb0-9093-55bb866ec721",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rv_FB_DP_distr = {\"TP\":[], \"FP\":[]}\n",
    "\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    depth = int(tag.split(\"_\")[-2])\n",
    "    freq = float(tag.split(\"_\")[-1])\n",
    "\n",
    "    # these are the INDELs that have been pre-filtered for non-L_GT and FB_found==1\n",
    "    INDEL_df = pd.read_csv(f\"{Rv_INDEL_source_dir}/{tag}.INDELs.csv\", low_memory=False)\n",
    "\n",
    "    TP_df = INDEL_df[INDEL_df.GT == 1]\n",
    "    FP_df = INDEL_df[INDEL_df.GT == 0]\n",
    "\n",
    "    Rv_FB_DP_distr[\"TP\"] += list(TP_df[\"FB_DP\"].values/depth)\n",
    "    Rv_FB_DP_distr[\"FP\"] += list(FP_df[\"FB_DP\"].values/depth)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb3c41b0-9f25-4b1a-8feb-622d5ced51b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_H37Rv_FB_INDEL_DP_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(Rv_FB_DP_distr, out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2162a73b-0714-469a-95cf-abe72cc548e9",
   "metadata": {},
   "source": [
    "### Mapping quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26e2e566-ad06-482d-9d1d-08300e261b0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rv_FB_MQ_distr = {\"TP\":[], \"FP\":[]}\n",
    "\n",
    "df_i = 0\n",
    "\n",
    "count = 0\n",
    "for tag in tagsRv:\n",
    "\n",
    "    if count % 100 == 0:\n",
    "        print(count, end=\" \")\n",
    "    count += 1\n",
    "\n",
    "    depth = int(tag.split(\"_\")[-2])\n",
    "    freq = float(tag.split(\"_\")[-1])\n",
    "\n",
    "    # these are the INDELs that have been pre-filtered for non-L_GT and FB_found==1\n",
    "    INDEL_df = pd.read_csv(f\"{Rv_INDEL_source_dir}/{tag}.INDELs.csv\", low_memory=False)\n",
    "\n",
    "    TP_df = INDEL_df[INDEL_df.GT == 1]\n",
    "    FP_df = INDEL_df[INDEL_df.GT == 0]\n",
    "\n",
    "    Rv_FB_MQ_distr[\"TP\"] += list(TP_df[\"FB_MQ\"].values)\n",
    "    Rv_FB_MQ_distr[\"FP\"] += list(FP_df[\"FB_MQ\"].values)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "820fc3d8-4372-4e57-be09-c3ba1e8d31e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"./data/FP_characteristics/ISS_H37Rv_FB_INDEL_MQ_distribution.pkl\", \"wb\") as out_f:\n",
    "    pickle.dump(Rv_FB_MQ_distr, out_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b62d315d-6cc9-4ff2-84aa-83523f952062",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
